# Analyzing the data

## Checking species wise class distribution

In [1]:
import collections
speciesList=["Human","Cow","Cat","Mouse","Pig","Rhesus","Microbat","Elephant","Rabbit","Shrew"]
#a = [1,1,1,1,2,2,2,2,3,3,4,5,5]
def get_species_class_distribution(speciesName,inputList):
    indexOfSpecies=speciesList.index(speciesName)
    tempList=[]
    for i in inputList:
        tempList.append(i[indexOfSpecies])
    assert(len(tempList)==len(inputList))
    counter=collections.Counter(tempList)
    return counter

In [2]:
def get_species_wise_class_dist(LabelList):
    
    stringTable="<table><tr><th>Species</th><th>Class 0</th><th>Class 1</th><th>Class 2</th><th>Class 3</th><th>Class 4</th></tr>"
    for i in speciesList[1:]:
        counter=get_species_class_distribution(i,LabelList)
        stringTable+="<tr><td>"+i+"</td><td>"+str(round(counter[0]*100.0/len(LabelList),3))+"</td><td>"+str(round(counter[1]*100.0/len(LabelList),3))+"</td><td>"+str(round(counter[2]*100.0/len(LabelList),3))+"</td><td>"+str(round(counter[3]*100.0/len(LabelList),3))+"</td><td>"+str(round(counter[4]*100.0/len(LabelList),3))+"</td><td>"
    stringTable+="</table>"
    return stringTable

# Load the data

The below cells load data from 2 files
1. The training indices files which contains the indices of each species
2. The labels file which contain the output labels

In [3]:
import json
with open("TrainingIndices10Species_Updated_withAncestors_ModifiedDic.json", 'r') as f:
    trainingIndices = json.load(f)
with open("TrainingLabel10Species_Updated_WithAncestors_ModifiedDic.json", 'r') as f:
    trainingLabels = json.load(f)

In [4]:
print "Data Load Done"
print trainingIndices[0]
print trainingLabels[0]
print len(trainingIndices)
print len(trainingLabels)

Data Load Done
[u'3634915_Human.chr22_1', u'4133722_Cow.chr22.1_1', u'4189836_Cat.chr22.1_-1', u'-19', u'-19', u'3292135_Rhesus.chr22_-1', u'1889148_Microbat.chr22.2.1.1_1', u'2096506_Elephant.chr22_-1', u'2702055_Rabbit.chr22.2.1.1_1', u'-19']
[-1, 0, 0, 3, 3, 0, 0, 4, 0, 3]
10471799
10471799


We will add the human index in the label list as well so that we can map corresponding list in the indices with that of the label

In [5]:
for enum,i in enumerate(trainingLabels):
    indexStr=str(trainingIndices[enum][0])
    firstUnderscore=indexStr.index('_')
    intIndex=int(indexStr[0:firstUnderscore])
    i[0]=intIndex

In [6]:
print trainingIndices[0]
print trainingLabels[0]

[u'3634915_Human.chr22_1', u'4133722_Cow.chr22.1_1', u'4189836_Cat.chr22.1_-1', u'-19', u'-19', u'3292135_Rhesus.chr22_-1', u'1889148_Microbat.chr22.2.1.1_1', u'2096506_Elephant.chr22_-1', u'2702055_Rabbit.chr22.2.1.1_1', u'-19']
[3634915, 0, 0, 3, 3, 0, 0, 4, 0, 3]


Now that we can map X to y. We can desample the data to have equal classes for Cow. First lets check the class distribution before desampling.

In [7]:
currentDistribution = get_species_wise_class_dist(trainingLabels)

In [8]:
print currentDistribution

<table><tr><th>Species</th><th>Class 0</th><th>Class 1</th><th>Class 2</th><th>Class 3</th><th>Class 4</th></tr><tr><td>Cow</td><td>40.076</td><td>19.257</td><td>9.142</td><td>23.82</td><td>7.706</td><td><tr><td>Cat</td><td>53.187</td><td>14.717</td><td>8.718</td><td>15.56</td><td>7.819</td><td><tr><td>Mouse</td><td>26.723</td><td>24.444</td><td>8.155</td><td>34.087</td><td>6.592</td><td><tr><td>Pig</td><td>50.141</td><td>15.821</td><td>9.012</td><td>17.561</td><td>7.464</td><td><tr><td>Rhesus</td><td>78.394</td><td>5.488</td><td>5.678</td><td>6.136</td><td>4.303</td><td><tr><td>Microbat</td><td>49.576</td><td>16.676</td><td>8.914</td><td>17.777</td><td>7.056</td><td><tr><td>Elephant</td><td>51.389</td><td>15.801</td><td>8.707</td><td>16.754</td><td>7.349</td><td><tr><td>Rabbit</td><td>50.267</td><td>15.603</td><td>9.037</td><td>16.402</td><td>8.69</td><td><tr><td>Shrew</td><td>24.88</td><td>24.961</td><td>8.156</td><td>35.899</td><td>6.105</td><td></table>


<table><tr><th>Species</th><th>Class 0</th><th>Class 1</th><th>Class 2</th><th>Class 3</th><th>Class 4</th></tr><tr><td>Cow</td><td>40.076</td><td>19.257</td><td>9.142</td><td>23.82</td><td>7.706</td><td><tr><td>Cat</td><td>53.187</td><td>14.717</td><td>8.718</td><td>15.56</td><td>7.819</td><td><tr><td>Mouse</td><td>26.723</td><td>24.444</td><td>8.155</td><td>34.087</td><td>6.592</td><td><tr><td>Pig</td><td>50.141</td><td>15.821</td><td>9.012</td><td>17.561</td><td>7.464</td><td><tr><td>Rhesus</td><td>78.394</td><td>5.488</td><td>5.678</td><td>6.136</td><td>4.303</td><td><tr><td>Microbat</td><td>49.576</td><td>16.676</td><td>8.914</td><td>17.777</td><td>7.056</td><td><tr><td>Elephant</td><td>51.389</td><td>15.801</td><td>8.707</td><td>16.754</td><td>7.349</td><td><tr><td>Rabbit</td><td>50.267</td><td>15.603</td><td>9.037</td><td>16.402</td><td>8.69</td><td><tr><td>Shrew</td><td>24.88</td><td>24.961</td><td>8.156</td><td>35.899</td><td>6.105</td><td></table>

Now we can desample the label list to contain equal class distribution of cow species. First let's calculate the total number of records in the minimum class,i.e. class 4 

### Desampling the data

In [9]:
minNumberOfRecords = (7.706/100)*len(trainingLabels)
print minNumberOfRecords

806956.83094


In [10]:
DesampledLabelList =[]
count0,count1,count2,count3,count4 =0,0,0,0,0
for ind,i in enumerate(trainingLabels):
    if(i[1] == 2):
        count2+=1
        if(count2<minNumberOfRecords):
            DesampledLabelList.append(i)
    if(i[1] == 0):
        count0+=1
        if(count0<minNumberOfRecords):
            DesampledLabelList.append(i)
    if(i[1] == 1):
        count1+=1
        if(count1<minNumberOfRecords):
            DesampledLabelList.append(i)
    if(i[1] == 3):
        count3+=1
        if(count3<minNumberOfRecords):
            DesampledLabelList.append(i)
    if(i[1] == 4):
        count4+=1
        if(count4<minNumberOfRecords):
            DesampledLabelList.append(i)
print len(DesampledLabelList)
    

4034749


In [11]:
newDistribution = get_species_wise_class_dist(DesampledLabelList)

In [12]:
print newDistribution
print DesampledLabelList[0]
print trainingIndices[0]

<table><tr><th>Species</th><th>Class 0</th><th>Class 1</th><th>Class 2</th><th>Class 3</th><th>Class 4</th></tr><tr><td>Cow</td><td>20.0</td><td>20.0</td><td>20.0</td><td>20.0</td><td>19.999</td><td><tr><td>Cat</td><td>44.846</td><td>16.363</td><td>14.545</td><td>15.065</td><td>9.18</td><td><tr><td>Mouse</td><td>22.405</td><td>25.695</td><td>12.717</td><td>32.09</td><td>7.093</td><td><tr><td>Pig</td><td>40.571</td><td>17.637</td><td>15.313</td><td>16.919</td><td>9.56</td><td><tr><td>Rhesus</td><td>71.657</td><td>7.303</td><td>9.322</td><td>6.235</td><td>5.483</td><td><tr><td>Microbat</td><td>41.63</td><td>18.172</td><td>14.675</td><td>17.157</td><td>8.366</td><td><tr><td>Elephant</td><td>44.187</td><td>17.703</td><td>14.079</td><td>16.063</td><td>7.969</td><td><tr><td>Rabbit</td><td>44.098</td><td>16.467</td><td>14.038</td><td>15.857</td><td>9.541</td><td><tr><td>Shrew</td><td>20.142</td><td>26.811</td><td>13.064</td><td>33.566</td><td>6.416</td><td></table>
[3634915, 0, 0, 3, 3, 0, 0,

<table><tr><th>Species</th><th>Class 0</th><th>Class 1</th><th>Class 2</th><th>Class 3</th><th>Class 4</th></tr><tr><td>Cow</td><td>20.0</td><td>20.0</td><td>20.0</td><td>20.0</td><td>19.999</td><td><tr><td>Cat</td><td>44.846</td><td>16.363</td><td>14.545</td><td>15.065</td><td>9.18</td><td><tr><td>Mouse</td><td>22.405</td><td>25.695</td><td>12.717</td><td>32.09</td><td>7.093</td><td><tr><td>Pig</td><td>40.571</td><td>17.637</td><td>15.313</td><td>16.919</td><td>9.56</td><td><tr><td>Rhesus</td><td>71.657</td><td>7.303</td><td>9.322</td><td>6.235</td><td>5.483</td><td><tr><td>Microbat</td><td>41.63</td><td>18.172</td><td>14.675</td><td>17.157</td><td>8.366</td><td><tr><td>Elephant</td><td>44.187</td><td>17.703</td><td>14.079</td><td>16.063</td><td>7.969</td><td><tr><td>Rabbit</td><td>44.098</td><td>16.467</td><td>14.038</td><td>15.857</td><td>9.541</td><td><tr><td>Shrew</td><td>20.142</td><td>26.811</td><td>13.064</td><td>33.566</td><td>6.416</td><td></table>

In [13]:
del trainingLabels# this list is no longer required we have the desampled list

We are going to fetch the nucleotides for the indices in trainingIndices list but we need to keep the human index intact even after getting the nucleotides, so lets add an additional human index to the trainingIndices list

In [14]:
for enum,i in enumerate(trainingIndices):
    indexStr=str(i[0])
    firstUnderscore=indexStr.index('_')
    intIndex=int(indexStr[0:firstUnderscore])
    i.append(intIndex)

In [15]:
print trainingIndices[0]
print trainingIndices[0][0:10]

[u'3634915_Human.chr22_1', u'4133722_Cow.chr22.1_1', u'4189836_Cat.chr22.1_-1', u'-19', u'-19', u'3292135_Rhesus.chr22_-1', u'1889148_Microbat.chr22.2.1.1_1', u'2096506_Elephant.chr22_-1', u'2702055_Rabbit.chr22.2.1.1_1', u'-19', 3634915]
[u'3634915_Human.chr22_1', u'4133722_Cow.chr22.1_1', u'4189836_Cat.chr22.1_-1', u'-19', u'-19', u'3292135_Rhesus.chr22_-1', u'1889148_Microbat.chr22.2.1.1_1', u'2096506_Elephant.chr22_-1', u'2702055_Rabbit.chr22.2.1.1_1', u'-19']


## Getting nucleotides from indices

Now that we have the trainingIndices list; we can proceed to get the nucleotides from the fasta files of those species and replace the indices with the appropriate nucleotides. First we will load all the fastas for the species in their specific lists

In [17]:
from Bio import SeqIO
humanSeqName=[]
humanSeqFasta=[]
cowSeqName=[]
cowSeqFasta=[]
mouseSeqName=[]
mouseSeqFasta=[]
microbatSeqName=[]
microbatSeqFasta=[]
pigSeqName=[]
pigSeqFasta=[]
rhesusSeqName=[]
rhesusSeqFasta=[]
rabbitSeqName=[]
rabbitSeqFasta=[]
shrewSeqName=[]
shrewSeqFasta=[]
elephantSeqName=[]
elephantSeqFasta=[]
catSeqName=[]
catSeqFasta=[]
for seq_record in SeqIO.parse("importantFiles/Human", "fasta"):
        humanSeqName.append(seq_record.id)
        humanSeqFasta.append(str(seq_record.seq))
for seq_record in SeqIO.parse("importantFiles/Cow", "fasta"):
        cowSeqName.append(seq_record.id)
        cowSeqFasta.append(seq_record.seq)
for seq_record in SeqIO.parse("importantFiles/Mouse", "fasta"):
        mouseSeqName.append(seq_record.id)
        mouseSeqFasta.append(seq_record.seq)
for seq_record in SeqIO.parse("importantFiles/Microbat", "fasta"):
        microbatSeqName.append(seq_record.id)
        microbatSeqFasta.append(seq_record.seq)
for seq_record in SeqIO.parse("importantFiles/Pig", "fasta"):
        pigSeqName.append(seq_record.id)
        pigSeqFasta.append(seq_record.seq)
for seq_record in SeqIO.parse("importantFiles/Rhesus", "fasta"):
        rhesusSeqName.append(seq_record.id)
        rhesusSeqFasta.append(seq_record.seq)
for seq_record in SeqIO.parse("importantFiles/Rabbit", "fasta"):
        rabbitSeqName.append(seq_record.id)
        rabbitSeqFasta.append(seq_record.seq)
for seq_record in SeqIO.parse("importantFiles/Shrew", "fasta"):
        shrewSeqName.append(seq_record.id)
        shrewSeqFasta.append(seq_record.seq)
for seq_record in SeqIO.parse("importantFiles/Elephant", "fasta"):
        elephantSeqName.append(seq_record.id)
        elephantSeqFasta.append(seq_record.seq)
for seq_record in SeqIO.parse("importantFiles/Cat", "fasta"):
        catSeqName.append(seq_record.id)
        catSeqFasta.append(seq_record.seq)

In Case of negative strand we need the complementary base pairs. Creating the function which returns the complementary base pair of a nucleotide.

In [18]:
def getComplementBasePair(inputBasePair):
    if inputBasePair=='A':
        return 'T'
    elif inputBasePair=='T':
        return 'A'
    elif inputBasePair=='C':
        return 'G'
    elif inputBasePair=='G':
        return 'C'

The below cell processes the combinedList and using the nucleotide list of each species replaces the combined list with nucleotides. At the end of the cell, you have a combined list with nucleotides.

In [19]:
#["Cow","Cat","Mouse","Pig","Rhesus","Microbat","Elephant","Rabbit","Shrew"]
print "total length:",len(trainingIndices)
#trainingNucleotide=[]
for enum,i in enumerate(trainingIndices):
    if enum%10000==0:
        print enum," done"
    temp=i[0:10]
    for index,k in enumerate(temp):
        #print k
        if index%10==0:
            if k==-19 or k=="-19":
                humanNucleotide='-'
            else:
                indexStr = str(k)
                firstUnderscore= indexStr.index('_')
                secondUnderscore= indexStr.rfind('_')
                seqNameInIndex=indexStr[firstUnderscore+1:secondUnderscore]
                intIndex=int(indexStr[0:firstUnderscore])
                strand=int(indexStr[secondUnderscore+1:])
                if strand==1:
                    humanNucleotide=humanSeqFasta[humanSeqName.index(seqNameInIndex)][intIndex].upper()
                else:
                    posNucleotide=humanSeqFasta[humanSeqName.index(seqNameInIndex)][intIndex].upper()
                    humanNucleotide=getComplementBasePair(posNucleotide)
            trainingIndices[enum][index]=humanNucleotide
            #temp.append(humanNucleotide)
        if index%10==1:
            try:
                if k==-19 or k=="-19":
                    cowNucleotide='-'
                else:
                    indexStr = str(k)
                    firstUnderscore= indexStr.index('_')
                    secondUnderscore= indexStr.rfind('_')
                    seqNameInIndex=indexStr[firstUnderscore+1:secondUnderscore]
                    intIndex=int(indexStr[0:firstUnderscore])
                    strand=int(indexStr[secondUnderscore+1:])
                    if strand==1:
                        cowNucleotide=cowSeqFasta[cowSeqName.index(seqNameInIndex)][intIndex].upper()
                    else:
                        posNucleotide=cowSeqFasta[cowSeqName.index(seqNameInIndex)][intIndex].upper()
                        cowNucleotide=getComplementBasePair(posNucleotide)
                trainingIndices[enum][index]=cowNucleotide
            except:
                continue
            #temp.append(cowNucleotide)
        if index%10==2:
            if k==-19 or k=="-19":
                catNucleotide='-'
            else:
                indexStr = str(k)
                firstUnderscore= indexStr.index('_')
                secondUnderscore= indexStr.rfind('_')
                seqNameInIndex=indexStr[firstUnderscore+1:secondUnderscore]
                intIndex=int(indexStr[0:firstUnderscore])
                strand=int(indexStr[secondUnderscore+1:])
                if strand==1:
                    catNucleotide=catSeqFasta[catSeqName.index(seqNameInIndex)][intIndex].upper()
                else:
                    posNucleotide=catSeqFasta[catSeqName.index(seqNameInIndex)][intIndex].upper()
                    catNucleotide=getComplementBasePair(posNucleotide)
            trainingIndices[enum][index]=catNucleotide
            #temp.append(catNucleotide)
        if index%10==3:
            if k==-19 or k=="-19":
                mouseNucleotide='-'
            else:
                indexStr = str(k)
                firstUnderscore= indexStr.index('_')
                secondUnderscore= indexStr.rfind('_')
                seqNameInIndex=indexStr[firstUnderscore+1:secondUnderscore]
                intIndex=int(indexStr[0:firstUnderscore])
                strand=int(indexStr[secondUnderscore+1:])
                if strand==1:
                    mouseNucleotide=mouseSeqFasta[mouseSeqName.index(seqNameInIndex)][intIndex].upper()
                else:
                    posNucleotide=mouseSeqFasta[mouseSeqName.index(seqNameInIndex)][intIndex].upper()
                    mouseNucleotide=getComplementBasePair(posNucleotide)
            trainingIndices[enum][index]=mouseNucleotide
            #temp.append(mouseNucleotide)
        if index%10==4:
            if k==-19 or k=="-19":
                pigNucleotide='-'
            else:
                indexStr = str(k)
                firstUnderscore= indexStr.index('_')
                secondUnderscore= indexStr.rfind('_')
                seqNameInIndex=indexStr[firstUnderscore+1:secondUnderscore]
                intIndex=int(indexStr[0:firstUnderscore])
                strand=int(indexStr[secondUnderscore+1:])
                if strand==1:
                    pigNucleotide=pigSeqFasta[pigSeqName.index(seqNameInIndex)][intIndex].upper()
                else:
                    posNucleotide=pigSeqFasta[pigSeqName.index(seqNameInIndex)][intIndex].upper()
                    pigNucleotide=getComplementBasePair(posNucleotide)
            trainingIndices[enum][index]=pigNucleotide
            #temp.append(pigNucleotide)
        if index%10==5:
            if k==-19 or k=="-19":
                rhesusNucleotide='-'
            else:
                indexStr = str(k)
                firstUnderscore= indexStr.index('_')
                secondUnderscore= indexStr.rfind('_')
                seqNameInIndex=indexStr[firstUnderscore+1:secondUnderscore]
                intIndex=int(indexStr[0:firstUnderscore])
                strand=int(indexStr[secondUnderscore+1:])
                if strand==1:
                    rhesusNucleotide=rhesusSeqFasta[rhesusSeqName.index(seqNameInIndex)][intIndex].upper()
                else:
                    posNucleotide=rhesusSeqFasta[rhesusSeqName.index(seqNameInIndex)][intIndex].upper()
                    rhesusNucleotide=getComplementBasePair(posNucleotide)
            trainingIndices[enum][index]=rhesusNucleotide
        if index%10==6:
            if k==-19 or k=="-19":
                microbatNucleotide='-'
            else:
                indexStr = str(k)
                firstUnderscore= indexStr.index('_')
                secondUnderscore= indexStr.rfind('_')
                seqNameInIndex=indexStr[firstUnderscore+1:secondUnderscore]
                intIndex=int(indexStr[0:firstUnderscore])
                strand=int(indexStr[secondUnderscore+1:])
                if strand==1:
                    microbatNucleotide=microbatSeqFasta[microbatSeqName.index(seqNameInIndex)][intIndex].upper()
                else:
                    posNucleotide=microbatSeqFasta[microbatSeqName.index(seqNameInIndex)][intIndex].upper()
                    microbatNucleotide=getComplementBasePair(posNucleotide)
            trainingIndices[enum][index]=microbatNucleotide
        if index%10==7:
            if k==-19 or k=="-19":
                elephantNucleotide='-'
            else:
                indexStr = str(k)
                firstUnderscore= indexStr.index('_')
                secondUnderscore= indexStr.rfind('_')
                seqNameInIndex=indexStr[firstUnderscore+1:secondUnderscore]
                intIndex=int(indexStr[0:firstUnderscore])
                strand=int(indexStr[secondUnderscore+1:])
                if strand==1:
                    elephantNucleotide=elephantSeqFasta[elephantSeqName.index(seqNameInIndex)][intIndex].upper()
                else:
                    posNucleotide=elephantSeqFasta[elephantSeqName.index(seqNameInIndex)][intIndex].upper()
                    elephantNucleotide=getComplementBasePair(posNucleotide)
            trainingIndices[enum][index]=elephantNucleotide
        if index%10==8:
            if k==-19 or k=="-19":
                rabbitNucleotide='-'
            else:
                indexStr = str(k)
                firstUnderscore= indexStr.index('_')
                secondUnderscore= indexStr.rfind('_')
                seqNameInIndex=indexStr[firstUnderscore+1:secondUnderscore]
                intIndex=int(indexStr[0:firstUnderscore])
                strand=int(indexStr[secondUnderscore+1:])
                if strand==1:
                    rabbitNucleotide=rabbitSeqFasta[rabbitSeqName.index(seqNameInIndex)][intIndex].upper()
                else:
                    posNucleotide=rabbitSeqFasta[rabbitSeqName.index(seqNameInIndex)][intIndex].upper()
                    rabbitNucleotide=getComplementBasePair(posNucleotide)
            trainingIndices[enum][index]=rabbitNucleotide
        if index%10==9:
            if k==-19 or k=="-19":
                shrewNucleotide='-'
            else:
                indexStr = str(k)
                firstUnderscore= indexStr.index('_')
                secondUnderscore= indexStr.rfind('_')
                seqNameInIndex=indexStr[firstUnderscore+1:secondUnderscore]
                intIndex=int(indexStr[0:firstUnderscore])
                strand=int(indexStr[secondUnderscore+1:])
                if strand==1:
                    shrewNucleotide=shrewSeqFasta[shrewSeqName.index(seqNameInIndex)][intIndex].upper()
                else:
                    posNucleotide=shrewSeqFasta[shrewSeqName.index(seqNameInIndex)][intIndex].upper()
                    shrewNucleotide=getComplementBasePair(posNucleotide)
            trainingIndices[enum][index]=shrewNucleotide
print "Done"

total length: 10471799
0  done
10000  done
20000  done
30000  done
40000  done
50000  done
60000  done
70000  done
80000  done
90000  done
100000  done
110000  done
120000  done
130000  done
140000  done
150000  done
160000  done
170000  done
180000  done
190000  done
200000  done
210000  done
220000  done
230000  done
240000  done
250000  done
260000  done
270000  done
280000  done
290000  done
300000  done
310000  done
320000  done
330000  done
340000  done
350000  done
360000  done
370000  done
380000  done
390000  done
400000  done
410000  done
420000  done
430000  done
440000  done
450000  done
460000  done
470000  done
480000  done
490000  done
500000  done
510000  done
520000  done
530000  done
540000  done
550000  done
560000  done
570000  done
580000  done
590000  done
600000  done
610000  done
620000  done
630000  done
640000  done
650000  done
660000  done
670000  done
680000  done
690000  done
700000  done
710000  done
720000  done
730000  done
740000  done
750000  done
760

5920000  done
5930000  done
5940000  done
5950000  done
5960000  done
5970000  done
5980000  done
5990000  done
6000000  done
6010000  done
6020000  done
6030000  done
6040000  done
6050000  done
6060000  done
6070000  done
6080000  done
6090000  done
6100000  done
6110000  done
6120000  done
6130000  done
6140000  done
6150000  done
6160000  done
6170000  done
6180000  done
6190000  done
6200000  done
6210000  done
6220000  done
6230000  done
6240000  done
6250000  done
6260000  done
6270000  done
6280000  done
6290000  done
6300000  done
6310000  done
6320000  done
6330000  done
6340000  done
6350000  done
6360000  done
6370000  done
6380000  done
6390000  done
6400000  done
6410000  done
6420000  done
6430000  done
6440000  done
6450000  done
6460000  done
6470000  done
6480000  done
6490000  done
6500000  done
6510000  done
6520000  done
6530000  done
6540000  done
6550000  done
6560000  done
6570000  done
6580000  done
6590000  done
6600000  done
6610000  done
6620000  done
663000

In [20]:
print len(trainingIndices)
print trainingIndices[0]
for i in trainingIndices:
    for j in i[0:10]:
        if('_' in j):
            print i

10471799
['C', 'C', 'A', '-', '-', 'A', 'C', 'A', 'C', '-', 3634915]
['C', u'2652956_Cow.chr22.2.1.1.1_-1', 'C', '-', 'T', 'C', 'A', '-', 'C', '-', 4276219]


In [21]:
for indx,i in enumerate(trainingIndices):
    if(i[1] =='2652956_Cow.chr22.2.1.1.1_-1'):
        print 'hello'
        trainingIndices.remove(i)
print len(trainingIndices)

hello
10471798


In [22]:
print trainingIndices[0]#checking the data format. the 10th index is human index
print trainingIndices[1]

['C', 'C', 'A', '-', '-', 'A', 'C', 'A', 'C', '-', 3634915]
['G', 'G', 'G', 'G', 'G', 'G', 'A', 'G', 'G', '-', 7224579]


Creating a dictionary with the human index as key and everything else as value

In [23]:
trainingIndicesDic={}
for i in trainingIndices:
    trainingIndicesDic[i[10]]=i[0:10]

In [24]:
assert(len(trainingIndicesDic)==len(trainingIndices))

## Match Mismatch bits

We create another list of features that are for match mismatch bits w.r.t. to humans. if the nucleotide matches with respect to human then set match bit to 1. else 0.

In [25]:
matchMismatchBit = []
for i in trainingIndices:
    temp = []
    humanNucleotide = i[0]
    for j in i[1:10]:
        if j==humanNucleotide:
            temp.append(1)
        else:
            temp.append(0)
    temp.append(i[10])
    matchMismatchBit.append(temp)
assert(len(matchMismatchBit)==len(trainingIndices))

In [26]:
print len(matchMismatchBit)
print len(trainingIndices)
print matchMismatchBit[0]
print trainingIndices[0]

10471798
10471798
[1, 0, 0, 0, 0, 1, 0, 1, 0, 3634915]
['C', 'C', 'A', '-', '-', 'A', 'C', 'A', 'C', '-', 3634915]


In [27]:
matchMismatchDic={}
for i in matchMismatchBit:
    matchMismatchDic[i[9]]=i[0:9]

In [28]:
del trainingIndices

In [29]:
del matchMismatchBit

## Ancestor Sequences

In [31]:
AncestorNucleotide=[]
with open("TrainingIndices10Species_onlyHumanwithAncestors.json", 'r') as f:
    AncestorNucleotide = json.load(f)

In [32]:
print AncestorNucleotide[0]#checking the data format loaded list

[u'0_Human.chr22_1', [u'G', u'G', u'-', u'G', u'G', u'G', u'G', u'G', u'G']]


In [33]:
AncestorSequenceList=[]
for i in AncestorNucleotide:
    indexStr=str(i[0])
    firstUnderscore=indexStr.index('_')
    intIndex=int(indexStr[0:firstUnderscore])
    i[1].append(intIndex)
    AncestorSequenceList.append(i[1])

In [34]:
assert(len(AncestorSequenceList)==len(AncestorNucleotide))
print AncestorSequenceList[0:5]

[[u'G', u'G', u'-', u'G', u'G', u'G', u'G', u'G', u'G', 0], [u'A', u'A', u'A', u'A', u'A', u'A', u'A', u'A', u'A', 10000000], [u'A', u'A', u'A', u'A', u'A', u'A', u'A', u'A', u'A', 10000001], [u'A', u'A', u'A', u'A', u'A', u'A', u'A', u'A', u'A', 10000002], [u'A', u'A', u'A', u'A', u'A', u'A', u'A', u'A', u'A', 10000003]]


In [35]:
del AncestorNucleotide

## Ancestor Transition Transversion bits

For each branch in the phylogeny tree, we would see if their is a transition/transversion/indel along that branch going from the ancestor of that branch to the descendant. We have used 6 bits for encoding, 1 bit for indicating transition, 1 bit for indicating transversion,1 bit for indicating insertion, 1 bit for indicating deletion, 1 bit if no change in the branch, 1 bit if both the nucleotides are gaps. The encoding scheme is defined below
#transitionbit,transversionbit,Match,Insertion,Deletion,Gap

<table>
<tr><td>Transition Bit</td><td>Transversion Bit</td><td>Same Nucleotide Bit</td><td>Insertion Bit</td><td>Deletion Bit</td><td>Gap Bit</td><td>Description</td></tr>
<tr><td>0</td><td>0</td><td>1</td><td>0</td><td>0</td><td>0</td><td>No change in that branch</td></tr>
<tr><td>1</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>Transition along that branch</td></tr>
<tr><td>0</td><td>1</td><td>0</td><td>0</td><td>0</td><td>0</td><td>Transversion along that branch</td></tr>
<tr><td>0</td><td>0</td><td>0</td><td>1</td><td>0</td><td>0</td><td>Insertion along that branch</td></tr>
<tr><td>0</td><td>0</td><td>0</td><td>0</td><td>1</td><td>0</td><td>Deletion along that branch</td></tr>
<tr><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>1</td><td>Both gaps along that branch</td></tr>
</table>


In [36]:
def get_transition_transversion_features(humanNucleotide,otherSpeciesNucleotide):
    #return format transitionbit,transversionbit,Match,Insertion,Deletion,Gap
    if (humanNucleotide==otherSpeciesNucleotide):
        return 0,0,1,0,0,0
    elif (otherSpeciesNucleotide=='-' and humanNucleotide=='-'):
        return 0,0,0,0,0,1
    elif (otherSpeciesNucleotide=='-' and humanNucleotide!='-'):
        return 0,0,0,1,0,0
    elif (otherSpeciesNucleotide!='-' and humanNucleotide=='-'):
        return 0,0,0,0,1,0
    elif ((humanNucleotide=='A' and otherSpeciesNucleotide=='G') or (humanNucleotide=='G' and otherSpeciesNucleotide=='A')):
        return 1,0,0,0,0,0
    elif ((humanNucleotide=='C' and otherSpeciesNucleotide=='T') or (humanNucleotide=='T' and otherSpeciesNucleotide=='C')):
        return 1,0,0,0,0,0
    else:
        return 0,1,0,0,0,0

Now that we have the ancestor data; we need to match each ancestor data with the corresponding human index and make a list for this feature in the same order in which the data is present in combined list. The combined list contains the human index in the 10th position. we will use that to make a connection between these 2 lists. Firstly we see that the human index in ancestor nucleotide contains the species and the chromosome name. Since we know that for human there is only 1 unique sequence name we will remove everything and keep only the index intact in this list.

In [37]:
print len(AncestorSequenceList[0])
print len(trainingIndicesDic)
print len(AncestorSequenceList)
print AncestorSequenceList[0]
print AncestorSequenceList[0][-1]
for i in range(9):
    print trainingIndicesDic[0][i]

10
10471798
10542728
[u'G', u'G', u'-', u'G', u'G', u'G', u'G', u'G', u'G', 0]
0
G
G
G
-
G
G
-
-
-


In [ ]:
#["Human","Cow","Cat","Mouse","Pig","Rhesus","Microbat","Elephant","Rabbit","Shrew"]
#[LC_P[i],LC_M_C_P[i],LM_R[i],LC_M[i],LC_M_C_P_S[i],LH_R_M_R[i],LH_R[i],LC_M_C_P_S_H_R_M_R[i],Lhg18[i]]
#format transitionbit,transversionbit,Match,Insertion,Deletion,Gap
AncestorTransitionTransversionBits = []
AncestorTransitionTransversionSum=[]
for i in AncestorSequenceList:
    if i[9] not in trainingIndicesDic:
        continue
    transitionSum,transversionSum,matchSum,insertionSum,delSum,gapSum=0,0,0,0,0,0
    temp = []
    #temp = i[0:20]
    #H - HR
    x,y,z,a,b,c = get_transition_transversion_features(trainingIndicesDic[i[9]][0],i[-4])
    temp.append(x)
    temp.append(y)
    temp.append(z)
    temp.append(a)
    temp.append(b)
    temp.append(c)
    if x==0 and y==0 and z==0 and a==0 and b==0 and c==1:
        gapSum+=1
    if x==0 and y==0 and z==0 and a==0 and b==1 and c==0:
        delSum+=1
    if x==0 and y==0 and z==0 and a==1 and b==0 and c==0:
        insertionSum+=1
    if x==0 and y==0 and z==1 and a==0 and b==0 and c==0:
        matchSum+=1
    if x==1 and y==0 and z==0 and a==0 and b==0 and c==0:
        transitionSum+=1
    if x==0 and y==1 and z==0 and a==0 and b==0 and c==0:
        transversionSum+=1
    #R - HR
    x,y,z,a,b,c = get_transition_transversion_features(trainingIndicesDic[i[9]][5],i[-4])
    temp.append(x)
    temp.append(y)
    temp.append(z)
    temp.append(a)
    temp.append(b)
    temp.append(c)
    if x==0 and y==0 and z==0 and a==0 and b==0 and c==1:
        gapSum+=1
    if x==0 and y==0 and z==0 and a==0 and b==1 and c==0:
        delSum+=1
    if x==0 and y==0 and z==0 and a==1 and b==0 and c==0:
        insertionSum+=1
    if x==0 and y==0 and z==1 and a==0 and b==0 and c==0:
        matchSum+=1
    if x==1 and y==0 and z==0 and a==0 and b==0 and c==0:
        transitionSum+=1
    if x==0 and y==1 and z==0 and a==0 and b==0 and c==0:
        transversionSum+=1
    #Cow - CowPig
    x,y,z,a,b,c = get_transition_transversion_features(trainingIndicesDic[i[9]][1],i[-10])
    temp.append(x)
    temp.append(y)
    temp.append(z)
    temp.append(a)
    temp.append(b)
    temp.append(c)
    if x==0 and y==0 and z==0 and a==0 and b==0 and c==1:
        gapSum+=1
    if x==0 and y==0 and z==0 and a==0 and b==1 and c==0:
        delSum+=1
    if x==0 and y==0 and z==0 and a==1 and b==0 and c==0:
        insertionSum+=1
    if x==0 and y==0 and z==1 and a==0 and b==0 and c==0:
        matchSum+=1
    if x==1 and y==0 and z==0 and a==0 and b==0 and c==0:
        transitionSum+=1
    if x==0 and y==1 and z==0 and a==0 and b==0 and c==0:
        transversionSum+=1
    #Pig - CowPig
    x,y,z,a,b,c = get_transition_transversion_features(trainingIndicesDic[i[9]][4],i[-10])
    temp.append(x)
    temp.append(y)
    temp.append(z)
    temp.append(a)
    temp.append(b)
    temp.append(c)
    if x==0 and y==0 and z==0 and a==0 and b==0 and c==1:
        gapSum+=1
    if x==0 and y==0 and z==0 and a==0 and b==1 and c==0:
        delSum+=1
    if x==0 and y==0 and z==0 and a==1 and b==0 and c==0:
        insertionSum+=1
    if x==0 and y==0 and z==1 and a==0 and b==0 and c==0:
        matchSum+=1
    if x==1 and y==0 and z==0 and a==0 and b==0 and c==0:
        transitionSum+=1
    if x==0 and y==1 and z==0 and a==0 and b==0 and c==0:
        transversionSum+=1
    #Cat - CatMicrobat
    x,y,z,a,b,c = get_transition_transversion_features(trainingIndicesDic[i[9]][2],i[-7])
    temp.append(x)
    temp.append(y)
    temp.append(z)
    temp.append(a)
    temp.append(b)
    temp.append(c)
    if x==0 and y==0 and z==0 and a==0 and b==0 and c==1:
        gapSum+=1
    if x==0 and y==0 and z==0 and a==0 and b==1 and c==0:
        delSum+=1
    if x==0 and y==0 and z==0 and a==1 and b==0 and c==0:
        insertionSum+=1
    if x==0 and y==0 and z==1 and a==0 and b==0 and c==0:
        matchSum+=1
    if x==1 and y==0 and z==0 and a==0 and b==0 and c==0:
        transitionSum+=1
    if x==0 and y==1 and z==0 and a==0 and b==0 and c==0:
        transversionSum+=1
    #Microbat - CatMicrobat
    x,y,z,a,b,c = get_transition_transversion_features(trainingIndicesDic[i[9]][6],i[-7])
    temp.append(x)
    temp.append(y)
    temp.append(z)
    temp.append(a)
    temp.append(b)
    temp.append(c)
    if x==0 and y==0 and z==0 and a==0 and b==0 and c==1:
        gapSum+=1
    if x==0 and y==0 and z==0 and a==0 and b==1 and c==0:
        delSum+=1
    if x==0 and y==0 and z==0 and a==1 and b==0 and c==0:
        insertionSum+=1
    if x==0 and y==0 and z==1 and a==0 and b==0 and c==0:
        matchSum+=1
    if x==1 and y==0 and z==0 and a==0 and b==0 and c==0:
        transitionSum+=1
    if x==0 and y==1 and z==0 and a==0 and b==0 and c==0:
        transversionSum+=1
    #Mouse - MouseRabbit
    x,y,z,a,b,c = get_transition_transversion_features(trainingIndicesDic[i[9]][3],i[-8])
    temp.append(x)
    temp.append(y)
    temp.append(z)
    temp.append(a)
    temp.append(b)
    temp.append(c)
    if x==0 and y==0 and z==0 and a==0 and b==0 and c==1:
        gapSum+=1
    if x==0 and y==0 and z==0 and a==0 and b==1 and c==0:
        delSum+=1
    if x==0 and y==0 and z==0 and a==1 and b==0 and c==0:
        insertionSum+=1
    if x==0 and y==0 and z==1 and a==0 and b==0 and c==0:
        matchSum+=1
    if x==1 and y==0 and z==0 and a==0 and b==0 and c==0:
        transitionSum+=1
    if x==0 and y==1 and z==0 and a==0 and b==0 and c==0:
        transversionSum+=1
    #Rabbit - MouseRabbit
    x,y,z,a,b,c = get_transition_transversion_features(trainingIndicesDic[i[9]][8],i[-9])
    temp.append(x)
    temp.append(y)
    temp.append(z)
    temp.append(a)
    temp.append(b)
    temp.append(c)
    if x==0 and y==0 and z==0 and a==0 and b==0 and c==1:
        gapSum+=1
    if x==0 and y==0 and z==0 and a==0 and b==1 and c==0:
        delSum+=1
    if x==0 and y==0 and z==0 and a==1 and b==0 and c==0:
        insertionSum+=1
    if x==0 and y==0 and z==1 and a==0 and b==0 and c==0:
        matchSum+=1
    if x==1 and y==0 and z==0 and a==0 and b==0 and c==0:
        transitionSum+=1
    if x==0 and y==1 and z==0 and a==0 and b==0 and c==0:
        transversionSum+=1
    #CM - CMCP
    x,y,z,a,b,c = get_transition_transversion_features(i[-7],i[-9])
    temp.append(x)
    temp.append(y)
    temp.append(z)
    temp.append(a)
    temp.append(b)
    temp.append(c)
    if x==0 and y==0 and z==0 and a==0 and b==0 and c==1:
        gapSum+=1
    if x==0 and y==0 and z==0 and a==0 and b==1 and c==0:
        delSum+=1
    if x==0 and y==0 and z==0 and a==1 and b==0 and c==0:
        insertionSum+=1
    if x==0 and y==0 and z==1 and a==0 and b==0 and c==0:
        matchSum+=1
    if x==1 and y==0 and z==0 and a==0 and b==0 and c==0:
        transitionSum+=1
    if x==0 and y==1 and z==0 and a==0 and b==0 and c==0:
        transversionSum+=1
    #CP - CMCP
    x,y,z,a,b,c = get_transition_transversion_features(i[-10],i[-9])
    temp.append(x)
    temp.append(y)
    temp.append(z)
    temp.append(a)
    temp.append(b)
    temp.append(c)
    if x==0 and y==0 and z==0 and a==0 and b==0 and c==1:
        gapSum+=1
    if x==0 and y==0 and z==0 and a==0 and b==1 and c==0:
        delSum+=1
    if x==0 and y==0 and z==0 and a==1 and b==0 and c==0:
        insertionSum+=1
    if x==0 and y==0 and z==1 and a==0 and b==0 and c==0:
        matchSum+=1
    if x==1 and y==0 and z==0 and a==0 and b==0 and c==0:
        transitionSum+=1
    if x==0 and y==1 and z==0 and a==0 and b==0 and c==0:
        transversionSum+=1
    #CMCP - CMCPS
    x,y,z,a,b,c = get_transition_transversion_features(i[-9],i[-6])
    temp.append(x)
    temp.append(y)
    temp.append(z)
    temp.append(a)
    temp.append(b)
    temp.append(c)
    if x==0 and y==0 and z==0 and a==0 and b==0 and c==1:
        gapSum+=1
    if x==0 and y==0 and z==0 and a==0 and b==1 and c==0:
        delSum+=1
    if x==0 and y==0 and z==0 and a==1 and b==0 and c==0:
        insertionSum+=1
    if x==0 and y==0 and z==1 and a==0 and b==0 and c==0:
        matchSum+=1
    if x==1 and y==0 and z==0 and a==0 and b==0 and c==0:
        transitionSum+=1
    if x==0 and y==1 and z==0 and a==0 and b==0 and c==0:
        transversionSum+=1
    #Shrew - CMCPS
    x,y,z,a,b,c = get_transition_transversion_features(trainingIndicesDic[i[9]][9],i[-6])
    temp.append(x)
    temp.append(y)
    temp.append(z)
    temp.append(a)
    temp.append(b)
    temp.append(c)
    if x==0 and y==0 and z==0 and a==0 and b==0 and c==1:
        gapSum+=1
    if x==0 and y==0 and z==0 and a==0 and b==1 and c==0:
        delSum+=1
    if x==0 and y==0 and z==0 and a==1 and b==0 and c==0:
        insertionSum+=1
    if x==0 and y==0 and z==1 and a==0 and b==0 and c==0:
        matchSum+=1
    if x==1 and y==0 and z==0 and a==0 and b==0 and c==0:
        transitionSum+=1
    if x==0 and y==1 and z==0 and a==0 and b==0 and c==0:
        transversionSum+=1
    #HR - HRMR
    x,y,z,a,b,c = get_transition_transversion_features(i[-4],i[-5])
    temp.append(x)
    temp.append(y)
    temp.append(z)
    temp.append(a)
    temp.append(b)
    temp.append(c)
    if x==0 and y==0 and z==0 and a==0 and b==0 and c==1:
        gapSum+=1
    if x==0 and y==0 and z==0 and a==0 and b==1 and c==0:
        delSum+=1
    if x==0 and y==0 and z==0 and a==1 and b==0 and c==0:
        insertionSum+=1
    if x==0 and y==0 and z==1 and a==0 and b==0 and c==0:
        matchSum+=1
    if x==1 and y==0 and z==0 and a==0 and b==0 and c==0:
        transitionSum+=1
    if x==0 and y==1 and z==0 and a==0 and b==0 and c==0:
        transversionSum+=1
    #MR - HRMR
    x,y,z,a,b,c = get_transition_transversion_features(i[-8],i[-5])
    temp.append(x)
    temp.append(y)
    temp.append(z)
    temp.append(a)
    temp.append(b)
    temp.append(c)
    if x==0 and y==0 and z==0 and a==0 and b==0 and c==1:
        gapSum+=1
    if x==0 and y==0 and z==0 and a==0 and b==1 and c==0:
        delSum+=1
    if x==0 and y==0 and z==0 and a==1 and b==0 and c==0:
        insertionSum+=1
    if x==0 and y==0 and z==1 and a==0 and b==0 and c==0:
        matchSum+=1
    if x==1 and y==0 and z==0 and a==0 and b==0 and c==0:
        transitionSum+=1
    if x==0 and y==1 and z==0 and a==0 and b==0 and c==0:
        transversionSum+=1
    #CMCPS - CMCPSHRMR
    x,y,z,a,b,c = get_transition_transversion_features(i[-6],i[-3])
    temp.append(x)
    temp.append(y)
    temp.append(z)
    temp.append(a)
    temp.append(b)
    temp.append(c)
    if x==0 and y==0 and z==0 and a==0 and b==0 and c==1:
        gapSum+=1
    if x==0 and y==0 and z==0 and a==0 and b==1 and c==0:
        delSum+=1
    if x==0 and y==0 and z==0 and a==1 and b==0 and c==0:
        insertionSum+=1
    if x==0 and y==0 and z==1 and a==0 and b==0 and c==0:
        matchSum+=1
    if x==1 and y==0 and z==0 and a==0 and b==0 and c==0:
        transitionSum+=1
    if x==0 and y==1 and z==0 and a==0 and b==0 and c==0:
        transversionSum+=1
    #HRMR - CMCPSHRMR
    x,y,z,a,b,c = get_transition_transversion_features(i[-5],i[-3])
    temp.append(x)
    temp.append(y)
    temp.append(z)
    temp.append(a)
    temp.append(b)
    temp.append(c)
    if x==0 and y==0 and z==0 and a==0 and b==0 and c==1:
        gapSum+=1
    if x==0 and y==0 and z==0 and a==0 and b==1 and c==0:
        delSum+=1
    if x==0 and y==0 and z==0 and a==1 and b==0 and c==0:
        insertionSum+=1
    if x==0 and y==0 and z==1 and a==0 and b==0 and c==0:
        matchSum+=1
    if x==1 and y==0 and z==0 and a==0 and b==0 and c==0:
        transitionSum+=1
    if x==0 and y==1 and z==0 and a==0 and b==0 and c==0:
        transversionSum+=1
    #CMCPSHRMR - Hg18
    x,y,z,a,b,c = get_transition_transversion_features(i[-3],i[-2])
    temp.append(x)
    temp.append(y)
    temp.append(z)
    temp.append(a)
    temp.append(b)
    temp.append(c)
    if x==0 and y==0 and z==0 and a==0 and b==0 and c==1:
        gapSum+=1
    if x==0 and y==0 and z==0 and a==0 and b==1 and c==0:
        delSum+=1
    if x==0 and y==0 and z==0 and a==1 and b==0 and c==0:
        insertionSum+=1
    if x==0 and y==0 and z==1 and a==0 and b==0 and c==0:
        matchSum+=1
    if x==1 and y==0 and z==0 and a==0 and b==0 and c==0:
        transitionSum+=1
    if x==0 and y==1 and z==0 and a==0 and b==0 and c==0:
        transversionSum+=1
    #Elephant - Hg18
    x,y,z,a,b,c = get_transition_transversion_features(trainingIndicesDic[i[9]][7],i[-2])
    temp.append(x)
    temp.append(y)
    temp.append(z)
    temp.append(a)
    temp.append(b)
    temp.append(c)
    if x==0 and y==0 and z==0 and a==0 and b==0 and c==1:
        gapSum+=1
    if x==0 and y==0 and z==0 and a==0 and b==1 and c==0:
        delSum+=1
    if x==0 and y==0 and z==0 and a==1 and b==0 and c==0:
        insertionSum+=1
    if x==0 and y==0 and z==1 and a==0 and b==0 and c==0:
        matchSum+=1
    if x==1 and y==0 and z==0 and a==0 and b==0 and c==0:
        transitionSum+=1
    if x==0 and y==1 and z==0 and a==0 and b==0 and c==0:
        transversionSum+=1
    temp.append(i[9])
    AncestorTransitionTransversionBits.append(temp)    
    AncestorTransitionTransversionSum.append([transitionSum,transversionSum,matchSum,insertionSum,delSum,gapSum,i[9]])

In [ ]:
print AncestorTransitionTransversionBits[1000]
print AncestorTransitionTransversionSum[1000]
print len(AncestorTransitionTransversionBits[0])
print len(AncestorTransitionTransversionSum[0])
print len(AncestorTransitionTransversionBits[1000][0:108])
print AncestorTransitionTransversionBits[1000][-1]

In [ ]:
AncestorTransitionTransversionBitsDic={}
for i in AncestorTransitionTransversionBits:
    AncestorTransitionTransversionBitsDic[i[-1]]=i[0:108]

In [ ]:
AncestorTransitionTransversionSumDic={}
for i in AncestorTransitionTransversionSum:
    AncestorTransitionTransversionSumDic[i[-1]]=i[0:6]

In [ ]:
AncestorSequenceListDic={}
for i in AncestorSequenceList:
    AncestorSequenceListDic[i[-1]]=i[0:9]

In [ ]:
DesampledLabelListDic={}
for i in DesampledLabelList:
    DesampledLabelListDic[i[0]]=i[1:]

In [ ]:
print trainingIndicesDic[0]
#print DesampledLabelListDic[2345]
print AncestorTransitionTransversionBitsDic[0]
print AncestorTransitionTransversionSumDic[0]
print matchMismatchDic[0]
print len(trainingIndicesDic[0])
#print len(DesampledLabelListDic[0])
print len(AncestorTransitionTransversionBitsDic[0])
print len(AncestorTransitionTransversionSumDic[0])
print len(matchMismatchDic[0])
print len(trainingIndicesDic)
print len(DesampledLabelListDic)
print len(AncestorTransitionTransversionBitsDic)
print len(AncestorTransitionTransversionSumDic)
print len(matchMismatchDic)

In [ ]:
del DesampledLabelList,AncestorSequenceList,AncestorTransitionTransversionBits,AncestorTransitionTransversionSum

In [ ]:
del humanSeqFasta,humanSeqName,rabbitSeqFasta,rabbitSeqName,rhesusSeqFasta,rhesusSeqName,pigSeqFasta,pigSeqName,catSeqFasta,catSeqName,cowSeqName,cowSeqFasta,shrewSeqFasta,shrewSeqName,elephantSeqFasta,elephantSeqName,microbatSeqFasta,microbatSeqName,mouseSeqFasta,mouseSeqName

To summarize we have the following dictionaries with the human index as key in each dictionary which containt the corresponding features.
<table>
<tr><th>List Name</th><th>Feature Description</th></tr>
<tr><td>trainingIndicesDic</td><td>The columns nucleotides</td></tr>
<tr><td>DesampledLabelListDic</td><td>The labels of the corresponding nucleotides</td></tr>
<tr><td>AncestorTransitionTransversionBitsDic</td><td>The transition transversion bits of every edge in the phylogeny tree</td></tr>
<tr><td>AncestorTransitionTransversionSumDic</td><td>The sum of transition transverstion bits of every edge in the tree</td></tr>
<tr><td>matchMismatchDic</td><td>The match mismatch bits of other species with respect to humans</td></tr>
</table>

## Feature Set 1(Only Sequence)

We will use the dictionary trainingIndicesDic and ofcourse the labels DesampledLabelListDic for this feature. We create the data set which can take the parameter window size, then ensure the class balance of the label set and then use a classifier to get the confusion matrices.

In [126]:
noOfNeighborsOnOneSide=0

In [173]:
TrainingX=[]
TrainingY=[]
print "Total Records:",len(DesampledLabelListDic)
for enum,key in enumerate(DesampledLabelListDic):
    if enum%100000==0 and enum!=0:
        print "Processed ",enum," records."
    keyDoesntExist = 0
    tempList=[]
    for j in reversed(range(1,(noOfNeighborsOnOneSide+1))):
        if key-j in trainingIndicesDic:
            tempList+=(trainingIndicesDic[key-j])
        else:
            keyDoesntExist = 1
            break
    if keyDoesntExist == 1:
        continue
    if key in trainingIndicesDic:
        tempList+=(trainingIndicesDic[key])
    else:
        continue
    for k in range(1,(noOfNeighborsOnOneSide+1)):
        if key+k in trainingIndicesDic:
            tempList+=(trainingIndicesDic[key+k])
        else:
            keyDoesntExist=1
            break
    if keyDoesntExist == 1:
        continue
    TrainingX.append(tempList)
    TrainingY.append(DesampledLabelListDic[key])

Total Records: 4034749
Processed  100000  records.
Processed  200000  records.
Processed  300000  records.
Processed  400000  records.
Processed  500000  records.
Processed  600000  records.
Processed  700000  records.
Processed  800000  records.
Processed  900000  records.
Processed  1000000  records.
Processed  1100000  records.
Processed  1200000  records.
Processed  1300000  records.
Processed  1400000  records.
Processed  1500000  records.
Processed  1600000  records.
Processed  1700000  records.
Processed  1800000  records.
Processed  1900000  records.
Processed  2000000  records.
Processed  2100000  records.
Processed  2200000  records.
Processed  2300000  records.
Processed  2400000  records.
Processed  2500000  records.
Processed  2600000  records.
Processed  2700000  records.
Processed  2800000  records.
Processed  2900000  records.
Processed  3000000  records.
Processed  3100000  records.
Processed  3200000  records.
Processed  3300000  records.
Processed  3400000  records.


In [174]:
print len(TrainingX)
print len(TrainingY)

4034748
4034748


## One Hot Encoding

We now need some kind of encoding to encode the nucleotides in the TrainingX. We will use one hot encoding here

In [175]:
print TrainingX[0]#print to check the format of data once

['T', 'T', 'T', 'T', 'T', 'T', 'T', 'T', 'T', 'T']


Defining a function for one hot encoding and getting IndicesOneHotEncoded list at the end.

In [176]:
def generate_one_hot_encoding(shapelist,NucleoHash):
    x=[]
    for count,z in enumerate(shapelist):
        if(count%100000==0):
            print count
        p = []
        for j,k in enumerate(z):
            if(is_number(k)):
                p.append(k)
            else:
                p+=NucleoHash[k]
        #p = [int(i) for i in p]
        x.append(p)
    return x

In [177]:
def is_number(s):
    try:
        float(s)
        return True
    except ValueError:
        pass
 
    try:
        import unicodedata
        unicodedata.numeric(s)
        return True
    except (TypeError, ValueError):
        pass
 
    return False

In [178]:
NucleoHash={'A':[1,0,0,0,0],'C':[0,1,0,0,0],'G':[0,0,1,0,0],'T':[0,0,0,1,0],'-':[0,0,0,0,1]}
#NucleoHash={'A':100,'C':101,'G':102,'T':103,'-':104}
TrainingXOneHotEncoded =[]
TrainingXOneHotEncoded = generate_one_hot_encoding(TrainingX,NucleoHash)

0
100000
200000
300000
400000
500000
600000
700000
800000
900000
1000000
1100000
1200000
1300000
1400000
1500000
1600000
1700000
1800000
1900000
2000000
2100000
2200000
2300000
2400000
2500000
2600000
2700000
2800000
2900000
3000000
3100000
3200000
3300000
3400000
3500000
3600000
3700000
3800000
3900000
4000000


In [184]:
#print IndicesList[0]
print TrainingXOneHotEncoded[0]
print len(TrainingXOneHotEncoded)
print len(TrainingY)
print TrainingY[0]

[0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0]
4034748
4034748
[0, 0, 0, 0, 0, 0, 0, 0, 0]


In [180]:
del TrainingX

## Sckit Learn Train-Test Split 

The below cell just splits the X_training and Y_training into 70:30 percent as training and testing data respectively. It also shuffles the data. 

In [188]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(TrainingXOneHotEncoded, TrainingY, test_size=0.30, random_state=42)

## Random Forest Classifier 

We will define the random forest classifier here and we will only use 1 million records for training and 300k for testing as we feel that amount of data is sufficient to select the proper features.

In [189]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(n_estimators=10,verbose=1,max_depth=10)

In [182]:
# from sklearn.tree import DecisionTreeClassifier
# clf = DecisionTreeClassifier(max_depth=9)

In [190]:
clf.fit(X_train,y_train)

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:1: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  """Entry point for launching an IPython kernel.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:   49.7s finished


RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=10, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=None, verbose=1,
            warm_start=False)

Now that the classifier is fit, lets predict for 300k records in test and then predict for the training data as well

In [191]:
y_pred=clf.predict(X_test)

[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    2.4s finished


In [192]:
y_train_pred=clf.predict(X_train)

[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    8.4s finished


In [207]:
print y_pred[0:10]
print y_test[0:10]

[3 1 4 0 4 2 0 4 0 2]
[[3], [1], [2], [2], [2], [0], [0], [2], [4], [2]]


## Finding Accuracy 

The below function outputs the accuracy when you input two lists the predicted and the other list.

In [193]:
def get_accuracy(yPred,yTest):
    if len(yPred)!=len(yTest):
        print "yPred and yTest have different dimensions."
        return
    overallCorrect=0
    stringTable="<table><tr><th>True_Predicted</th><th>Class 0</th><th>Class 1</th><th>Class 2</th><th>Class 3</th><th>Class 4</th><th>Percentage</th></tr>"
    #classxy read as true class x predicted class y
    class00,class01,class02,class03,class04=0,0,0,0,0
    class10,class11,class12,class13,class14=0,0,0,0,0
    class20,class21,class22,class23,class24=0,0,0,0,0
    class30,class31,class32,class33,class34=0,0,0,0,0
    class40,class41,class42,class43,class44=0,0,0,0,0
    for enum1,i in enumerate(yPred):
        for enum2,j in enumerate(i):
            if j==yTest[enum1][enum2]:
                overallCorrect+=1
            if yTest[enum1][enum2]==0:
                if j==0:
                    class00+=1
                if j==1:
                    class01+=1
                if j==2:
                    class02+=1
                if j==3:
                    class03+=1
                if j==4:
                    class04+=1
            if yTest[enum1][enum2]==1:
                if j==0:
                    class10+=1
                if j==1:
                    class11+=1
                if j==2:
                    class12+=1
                if j==3:
                    class13+=1
                if j==4:
                    class14+=1
            if yTest[enum1][enum2]==2:
                if j==0:
                    class20+=1
                if j==1:
                    class21+=1
                if j==2:
                    class22+=1
                if j==3:
                    class23+=1
                if j==4:
                    class24+=1
            if yTest[enum1][enum2]==3:
                if j==0:
                    class30+=1
                if j==1:
                    class31+=1
                if j==2:
                    class32+=1
                if j==3:
                    class33+=1
                if j==4:
                    class34+=1
            if yTest[enum1][enum2]==4:
                if j==0:
                    class40+=1
                if j==1:
                    class41+=1
                if j==2:
                    class42+=1
                if j==3:
                    class43+=1
                if j==4:
                    class44+=1
    accuracyPercentage=overallCorrect*1.0/(len(yPred[0])*len(yPred))
    print "Overall Accuracy is",accuracyPercentage
    print "Class 0:",class00,class01,class02,class03,class04,class00+class01+class02+class03+class04,class00*1.0/(class00+class01+class02+class03+class04)
    print "Class 1:",class10,class11,class12,class13,class14,class10+class11+class12+class13+class14,class11*1.0/(class10+class11+class12+class13+class14)
    print "Class 2:",class20,class21,class22,class23,class24,class20+class21+class22+class23+class24,class22*1.0/(class20+class21+class22+class23+class24)
    print "Class 3:",class30,class31,class32,class33,class34,class30+class31+class32+class33+class34,class33*1.0/(class30+class31+class32+class33+class34)
    print "Class 4:",class40,class41,class42,class43,class44,class40+class41+class42+class43+class44,class44*1.0/(class40+class41+class42+class43+class44)
    stringTable+="<tr><td>"+"Class 0"+"</td><td>"+str(class00)+"</td><td>"+str(class01)+"</td><td>"+str(class02)+"</td><td>"+str(class03)+"</td><td>"+str(class04)+"</td><td>"+str(round(class00*1.0/(class00+class01+class02+class03+class04),3))+"</td></tr>"
    stringTable+="<tr><td>"+"Class 1"+"</td><td>"+str(class10)+"</td><td>"+str(class11)+"</td><td>"+str(class12)+"</td><td>"+str(class13)+"</td><td>"+str(class14)+"</td><td>"+str(round(class11*1.0/(class10+class11+class12+class13+class14),3))+"</td></tr>"
    stringTable+="<tr><td>"+"Class 2"+"</td><td>"+str(class20)+"</td><td>"+str(class21)+"</td><td>"+str(class22)+"</td><td>"+str(class23)+"</td><td>"+str(class24)+"</td><td>"+str(round(class22*1.0/(class20+class21+class22+class23+class24),3))+"</td></tr>"
    stringTable+="<tr><td>"+"Class 3"+"</td><td>"+str(class30)+"</td><td>"+str(class31)+"</td><td>"+str(class32)+"</td><td>"+str(class33)+"</td><td>"+str(class34)+"</td><td>"+str(round(class33*1.0/(class30+class31+class32+class33+class34),3))+"</td></tr>"
    stringTable+="<tr><td>"+"Class 4"+"</td><td>"+str(class40)+"</td><td>"+str(class41)+"</td><td>"+str(class42)+"</td><td>"+str(class43)+"</td><td>"+str(class44)+"</td><td>"+str(round(class44*1.0/(class40+class41+class42+class43+class44),3))+"</td></tr>"
    stringTable+="<tr><td>Overall Accuracy</td><td></td><td></td><td></td><td></td><td></td><td>"+str(round(accuracyPercentage,3))+"</td></tr>"
    stringTable+="</table>"
    print stringTable

In [206]:
print "Training Accuracy"
get_accuracy(y_train_pred,y_train)
print "Test Accuracy"
get_accuracy(y_pred,y_test)

Test Accuracy
Overall Accuracy is 0.502561496995
Class 0: 127659 0 38202 0 76340 242201 0.527078748643
Class 1: 0 122921 0 119442 0 242363 0.507177250653
Class 2: 76736 0 92613 0 72719 242068 0.382590842243
Class 3: 0 88616 0 153489 0 242105 0.633976993453
Class 4: 88540 0 41517 0 111631 241688 0.461880606402
<table><tr><th>True_Predicted</th><th>Class 0</th><th>Class 1</th><th>Class 2</th><th>Class 3</th><th>Class 4</th><th>Percentage</th></tr><tr><td>Class 0</td><td>127659</td><td>0</td><td>38202</td><td>0</td><td>76340</td><td>0.527</td></tr><tr><td>Class 1</td><td>0</td><td>122921</td><td>0</td><td>119442</td><td>0</td><td>0.507</td></tr><tr><td>Class 2</td><td>76736</td><td>0</td><td>92613</td><td>0</td><td>72719</td><td>0.383</td></tr><tr><td>Class 3</td><td>0</td><td>88616</td><td>0</td><td>153489</td><td>0</td><td>0.634</td></tr><tr><td>Class 4</td><td>88540</td><td>0</td><td>41517</td><td>0</td><td>111631</td><td>0.462</td></tr><tr><td>Overall Accuracy</td><td></td><td></td><t


<table><tr><th>True_Predicted</th><th>Class 0</th><th>Class 1</th><th>Class 2</th><th>Class 3</th><th>Class 4</th><th>Percentage</th></tr><tr><td>Class 0</td><td>9442428</td><td>4063</td><td>255871</td><td>2899</td><td>169271</td><td>0.956</td></tr><tr><td>Class 1</td><td>60366</td><td>2198705</td><td>2404</td><td>2432131</td><td>0</td><td>0.468</td></tr><tr><td>Class 2</td><td>2706103</td><td>4761</td><td>735012</td><td>1901</td><td>157996</td><td>0.204</td></tr><tr><td>Class 3</td><td>63029</td><td>1526481</td><td>604</td><td>3293692</td><td>0</td><td>0.674</td></tr><tr><td>Class 4</td><td>1947299</td><td>2625</td><td>164360</td><td>1736</td><td>245170</td><td>0.104</td></tr><tr><td>Overall Accuracy</td><td></td><td></td><td></td><td></td><td></td><td>0.626</td></tr></table>

<table><tr><th>True_Predicted</th><th>Class 0</th><th>Class 1</th><th>Class 2</th><th>Class 3</th><th>Class 4</th><th>Percentage</th></tr><tr><td>Class 0</td><td>4041033</td><td>1784</td><td>111051</td><td>1251</td><td>73227</td><td>0.956</td></tr><tr><td>Class 1</td><td>25785</td><td>937891</td><td>1013</td><td>1045475</td><td>0</td><td>0.467</td></tr><tr><td>Class 2</td><td>1164318</td><td>2080</td><td>313568</td><td>857</td><td>67910</td><td>0.202</td></tr><tr><td>Class 3</td><td>27311</td><td>657640</td><td>253</td><td>1409193</td><td>0</td><td>0.673</td></tr><tr><td>Class 4</td><td>834967</td><td>1202</td><td>70521</td><td>785</td><td>104710</td><td>0.103</td></tr><tr><td>Overall Accuracy</td><td></td><td></td><td></td><td></td><td></td><td>0.625</td></tr></table>

In [169]:
def get_accuracy_SpeciesWise(yPred,yTest,species):
    if len(yPred)!=len(yTest):
        print "yPred and yTest have different dimensions."
        return
    overallCorrect=0
    stringTable="<table><tr><th>species</th><th>Class 0</th><th>Class 1</th><th>Class 2</th><th>Class 3</th><th>Class 4</th><th>Percentage</th></tr>"
    #classxy read as true class x predicted class y
    class00,class01,class02,class03,class04=0,0,0,0,0
    class10,class11,class12,class13,class14=0,0,0,0,0
    class20,class21,class22,class23,class24=0,0,0,0,0
    class30,class31,class32,class33,class34=0,0,0,0,0
    class40,class41,class42,class43,class44=0,0,0,0,0
    speciesList=["Cow","Cat","Mouse","Pig","Rhesus","Microbat","Elephant","Rabbit","Shrew"]
    enum2 = speciesList.index(species)
    #print "enum2 is",enum2
    for enum1,i in enumerate(yPred):
        #print i
        j=i[enum2]
        j=int(j)
        #print j
        #print yTest[enum1][enum2]
        if j==yTest[enum1][enum2]:
            overallCorrect+=1
        if yTest[enum1][enum2]==0:
            if j==0:
                class00+=1
            if j==1:
                class01+=1
            if j==2:
                class02+=1
            if j==3:
                class03+=1
            if j==4:
                class04+=1
        if yTest[enum1][enum2]==1:
            if j==0:
                class10+=1
            if j==1:
                class11+=1
            if j==2:
                class12+=1
            if j==3:
                class13+=1
            if j==4:
                class14+=1
        if yTest[enum1][enum2]==2:
            if j==0:
                class20+=1
            if j==1:
                class21+=1
            if j==2:
                class22+=1
            if j==3:
                class23+=1
            if j==4:
                class24+=1
        if yTest[enum1][enum2]==3:
            if j==0:
                class30+=1
            if j==1:
                class31+=1
            if j==2:
                class32+=1
            if j==3:
                class33+=1
            if j==4:
                class34+=1
        if yTest[enum1][enum2]==4:
            if j==0:
                class40+=1
            if j==1:
                class41+=1
            if j==2:
                class42+=1
            if j==3:
                class43+=1
            if j==4:
                class44+=1
    accuracyPercentage=overallCorrect*1.0/(len(yPred))
    print "Overall Accuracy is",accuracyPercentage
    print "Class 0:",class00,class01,class02,class03,class04,class00+class01+class02+class03+class04,class00*1.0/(class00+class01+class02+class03+class04)
    print "Class 1:",class10,class11,class12,class13,class14,class10+class11+class12+class13+class14,class11*1.0/(class10+class11+class12+class13+class14)
    print "Class 2:",class20,class21,class22,class23,class24,class20+class21+class22+class23+class24,class22*1.0/(class20+class21+class22+class23+class24)
    print "Class 3:",class30,class31,class32,class33,class34,class30+class31+class32+class33+class34,class33*1.0/(class30+class31+class32+class33+class34)
    print "Class 4:",class40,class41,class42,class43,class44,class40+class41+class42+class43+class44,class44*1.0/(class40+class41+class42+class43+class44)
    stringTable+="<tr><td>"+"Class 0"+"</td><td>"+str(class00)+"</td><td>"+str(class01)+"</td><td>"+str(class02)+"</td><td>"+str(class03)+"</td><td>"+str(class04)+"</td><td>"+str(round(class00*1.0/(class00+class01+class02+class03+class04),3))+"</td></tr>"
    stringTable+="<tr><td>"+"Class 1"+"</td><td>"+str(class10)+"</td><td>"+str(class11)+"</td><td>"+str(class12)+"</td><td>"+str(class13)+"</td><td>"+str(class14)+"</td><td>"+str(round(class11*1.0/(class10+class11+class12+class13+class14),3))+"</td></tr>"
    stringTable+="<tr><td>"+"Class 2"+"</td><td>"+str(class20)+"</td><td>"+str(class21)+"</td><td>"+str(class22)+"</td><td>"+str(class23)+"</td><td>"+str(class24)+"</td><td>"+str(round(class22*1.0/(class20+class21+class22+class23+class24),3))+"</td></tr>"
    stringTable+="<tr><td>"+"Class 3"+"</td><td>"+str(class30)+"</td><td>"+str(class31)+"</td><td>"+str(class32)+"</td><td>"+str(class33)+"</td><td>"+str(class34)+"</td><td>"+str(round(class33*1.0/(class30+class31+class32+class33+class34),3))+"</td></tr>"
    stringTable+="<tr><td>"+"Class 4"+"</td><td>"+str(class40)+"</td><td>"+str(class41)+"</td><td>"+str(class42)+"</td><td>"+str(class43)+"</td><td>"+str(class44)+"</td><td>"+str(round(class44*1.0/(class40+class41+class42+class43+class44),3))+"</td></tr>"
    stringTable+="<tr><td>Overall Accuracy</td><td></td><td></td><td></td><td></td><td></td><td>"+str(round(accuracyPercentage,3))+"</td></tr>"
    stringTable+="</table>"
    print stringTable
    

In [170]:
print "Training Accuracy"
get_accuracy_SpeciesWise(y_train_pred,y_train,"Cow")
print "Test Accuracy"
get_accuracy_SpeciesWise(y_pred,y_test,"Cow")

Training Accuracy
Overall Accuracy is 0.500458339928
Class 0: 293477 24 101983 1 169270 564755 0.519653655125
Class 1: 0 293372 75 271146 0 564593 0.519616785897
Class 2: 180927 30 225936 0 157995 564888 0.399966010961
Class 3: 0 209323 22 355506 0 564851 0.629380137417
Class 4: 215248 32 104790 1 245165 565236 0.43373918151
<table><tr><th>species</th><th>Class 0</th><th>Class 1</th><th>Class 2</th><th>Class 3</th><th>Class 4</th><th>Percentage</th></tr><tr><td>Class 0</td><td>293477</td><td>24</td><td>101983</td><td>1</td><td>169270</td><td>0.52</td></tr><tr><td>Class 1</td><td>0</td><td>293372</td><td>75</td><td>271146</td><td>0</td><td>0.52</td></tr><tr><td>Class 2</td><td>180927</td><td>30</td><td>225936</td><td>0</td><td>157995</td><td>0.4</td></tr><tr><td>Class 3</td><td>0</td><td>209323</td><td>22</td><td>355506</td><td>0</td><td>0.629</td></tr><tr><td>Class 4</td><td>215248</td><td>32</td><td>104790</td><td>1</td><td>245165</td><td>0.434</td></tr><tr><td>Overall Accuracy</td><t

Training Accuracy
<table><tr><th>species</th><th>Class 0</th><th>Class 1</th><th>Class 2</th><th>Class 3</th><th>Class 4</th><th>Percentage</th></tr><tr><td>Class 0</td><td>293477</td><td>24</td><td>101983</td><td>1</td><td>169270</td><td>0.52</td></tr><tr><td>Class 1</td><td>0</td><td>293372</td><td>75</td><td>271146</td><td>0</td><td>0.52</td></tr><tr><td>Class 2</td><td>180927</td><td>30</td><td>225936</td><td>0</td><td>157995</td><td>0.4</td></tr><tr><td>Class 3</td><td>0</td><td>209323</td><td>22</td><td>355506</td><td>0</td><td>0.629</td></tr><tr><td>Class 4</td><td>215248</td><td>32</td><td>104790</td><td>1</td><td>245165</td><td>0.434</td></tr><tr><td>Overall Accuracy</td><td></td><td></td><td></td><td></td><td></td><td>0.5</td></tr></table>
Test Accuracy
<table><tr><th>species</th><th>Class 0</th><th>Class 1</th><th>Class 2</th><th>Class 3</th><th>Class 4</th><th>Percentage</th></tr><tr><td>Class 0</td><td>125289</td><td>15</td><td>43671</td><td>0</td><td>73226</td><td>0.517</td></tr><tr><td>Class 1</td><td>0</td><td>125179</td><td>24</td><td>117160</td><td>0</td><td>0.516</td></tr><tr><td>Class 2</td><td>77678</td><td>8</td><td>96473</td><td>0</td><td>67909</td><td>0.399</td></tr><tr><td>Class 3</td><td>0</td><td>90531</td><td>6</td><td>151568</td><td>0</td><td>0.626</td></tr><tr><td>Class 4</td><td>92035</td><td>12</td><td>44933</td><td>0</td><td>104708</td><td>0.433</td></tr><tr><td>Overall Accuracy</td><td></td><td></td><td></td><td></td><td></td><td>0.498</td></tr></table>

In [208]:
del TrainingXOneHotEncoded,TrainingY

## Unused Code

In [140]:
def generate_feature_set(shapelist,NucleoHash):
    x=[]
    for count,z in enumerate(shapelist):
        if(count%1000000==0):
            print count
        p = []
        for j,k in enumerate(z):
            if(is_number(k)):
                p.append(k)
            else:
                p.append(NucleoHash[k])
        x.append(p)
    return x

In [19]:
for enum,i in enumerate(AncestorSequenceList):
    indexStr=str(i[0])
    firstUnderscore=indexStr.index('_')
    intIndex=int(indexStr[0:firstUnderscore])
    i[0]=intIndex

Now we need to compare the corresponding human index in both lists, a good way to do this is to create a dictionary. We will create a dictionary on combined list. 
The key of the dictionary will be the human index and the value would we everthing in the combined list except the human index

In [20]:
a={}
for i in combinedList:
    temp = i[10]
    if (not a.has_key(temp)):
        a[temp] = i

In [21]:
print a[107]#check the data format
print len(a)

['G', 'G', 'G', '-', 'G', 'G', '-', 'A', '-', '-', 107, -1, 0, 0, 3, 0, 0, 3, 0, 3, 1]
10471798


We observe that there are cases where we have multiple data for same human index, and we don't want this inconsistency in our data so to check on it later, we'll create a dictionary of ancestors so that we can remove data which are multiple.

In [22]:
b={}
for i in AncestorNucleotide:
    temp = i[0]
    if (not b.has_key(temp)):
        b[temp] = i[1]

Let's just remove the duplicates, i will create a list back from the dictionary, this list will not have duplicates

In [23]:
AncestorNucleotide=[]
for j in b:
    AncestorNucleotide.append([j]+[b[j]])
print len(AncestorNucleotide)
assert(len(b)==len(AncestorNucleotide))

10476466


In [24]:
print b[0]
print AncestorNucleotide[0]

[u'G', u'G', u'-', u'G', u'G', u'G', u'G', u'G', u'G']
[0, [u'G', u'G', u'-', u'G', u'G', u'G', u'G', u'G', u'G']]


We will now create one list which contains the combination of both combined list and ancestor list

In [25]:
FinalCombinedList = []
for i in AncestorNucleotide:
    temp = i[0]
    if(a.has_key(temp) and b.has_key(temp)):
        templist = a[temp]+i[1]
        FinalCombinedList.append(templist)

In [26]:
print len(FinalCombinedList)
print len(combinedList)
print FinalCombinedList[2]
print combinedList[2]

10471798
10471798
['T', 'T', 'T', '-', 'T', 'T', '-', '-', '-', '-', 2, -1, 0, 0, 3, 0, 0, 3, 3, 3, 1, u'T', u'T', u'-', u'T', u'T', u'T', u'T', u'T', u'T']
['T', 'T', 'T', '-', 'T', 'T', '-', '-', '-', '-', 2, -1, 0, 0, 3, 0, 0, 3, 3, 3, 1]


Checking if the order of combinedList and FinalCombinedList is same. So that when we can later combine them for enhancing our feature.

In [27]:
for index,i in enumerate(FinalCombinedList):
    if i[10] != combinedList[index][10]:
        print "Warning"

Now that we have one list we just have to compare the nucleotides at each branch of the phylogeny tree to get the transition/transversion bits. First lets define a fuction which takes 2 nucleotides and outputs the transition/transversion bits. The below fucntion does that. it takes input as 2 nucleotide and returns 6 bits as defined above

Along with transition/transversion bits; a good idea would be to also get the sum of transitions;transversions;indels along each column. The below cell gets the bits along with their sum as well. We get 2 lists at the end;
1. AncestorTransitionTransversionBits - which contains the bits
2. AncestorTransitionTransversionSum - which contains the sum

In [32]:
print AncestorTransitionTransversionBits[0]
print AncestorTransitionTransversionSum[0]
print FinalCombinedList[0]
print combinedList[0]
print len(combinedList)
print len(AncestorTransitionTransversionBits)
print len(FinalCombinedList)

[0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0]
[0, 0, 14, 0, 4, 0]
['G', 'G', 'G', '-', 'G', 'G', '-', '-', '-', '-', 0, -1, 0, 0, 3, 0, 0, 3, 3, 3, 1, u'G', u'G', u'-', u'G', u'G', u'G', u'G', u'G', u'G']
['G', 'G', 'G', '-', 'G', 'G', '-', '-', '-', '-', 0, -1, 0, 0, 3, 0, 0, 3, 3, 3, 1]
10471798
10471798
10471798


In [33]:
for i in range(len(FinalCombinedList)):
    FinalCombinedList[i] = FinalCombinedList[i] + AncestorTransitionTransversionSum[i]
print FinalCombinedList[0]

['G', 'G', 'G', '-', 'G', 'G', '-', '-', '-', '-', 0, -1, 0, 0, 3, 0, 0, 3, 3, 3, 1, u'G', u'G', u'-', u'G', u'G', u'G', u'G', u'G', u'G', 0, 0, 14, 0, 4, 0]


Now we have the desired ancestor transition transversion feature set. We can delete all the intermediate lists which are no longer required to free up some memory

For each item in the indices list, appending the human index position at the end of each list. This is done so that we can sort the items based on human index because we need to find the flanking region of each index, and it would be easy to process that if the indices are sorted

In [6]:
for enum,i in enumerate(trainingIndices):
    indexStr=str(i[0])
    firstUnderscore=indexStr.index('_')
    intIndex=int(indexStr[0:firstUnderscore])
    i.append(intIndex)

Now creating one combined list that would contain the indices and the labels combined so that the labels don't get mapped to the wrong index while sorting. After combining sorting them based on the human index

In [7]:
combinedList=[]
for index,i in enumerate(range(len(trainingIndices))):
    combinedList.append(trainingIndices[index]+trainingLabels[index])
combinedList.sort(key=lambda x:x[10])

Since we have everything in the combined list we can delete the loaded training Indices and training labels.

In [8]:
del trainingIndices,trainingLabels

In [9]:
print combinedList[3]#print to see if data is loaded properly

[u'3_Human.chr22_1', u'3_Cow.chr22.1_1', u'3_Cat.chr22.1_1', u'-19', u'3349437_Pig.chr22.1.1+chr22.2.2+chr22.1.1+ch.._1', u'3_Rhesus.chr22_1', u'-19', u'-19', u'-19', u'-19', 3, -1, 0, 0, 3, 0, 0, 3, 3, 3, 1]


For now let's use a subset of the combined list; i.e. first million records. So we keep only the first million in memory

In [7]:
#combinedList = combinedList[0:1000000]

## Checking the range of continous human indices in combined list

Checking if the first million combined list items has a continuous human index of 1 to million; if not printing them into ranges of continous numbers

Let's start with storing the human indices of the first million records into a separate list

In [8]:
data=[]
for i in combinedList:
    data.append(i[10])
print len(data)

1000000


Now printing the ranges of the list named "data" 

In [9]:
from itertools import groupby
from operator import itemgetter
ranges = []
for key, group in groupby(enumerate(data), lambda (index, item): index - item):
    group = map(itemgetter(1), group)
    if len(group) > 1:
        ranges.append(xrange(group[0], group[-1]))
    else:
        ranges.append(group[0])

In [10]:
print data[7063]
print ranges

7063
[xrange(7063), xrange(7067, 9397), xrange(9721, 10734), xrange(11766, 17384), xrange(18206, 31722), xrange(31724, 38113), xrange(38129, 38217), xrange(38237, 39363), xrange(39395, 46283), xrange(46312, 53002), xrange(53007, 59781), xrange(59795, 62983), xrange(63039, 64486), xrange(64551, 78885), xrange(78924, 86076), xrange(86078, 86466), xrange(86558, 86767), xrange(86869, 87053), xrange(87455, 88479), xrange(88564, 90074), xrange(90115, 109407), xrange(109409, 109556), xrange(109589, 109822), xrange(109824, 115674), xrange(115686, 117449), xrange(117468, 118801), xrange(118804, 119474), xrange(119488, 122530), xrange(122597, 125730), xrange(125734, 150975), xrange(150996, 159133), xrange(159138, 162568), xrange(162612, 165337), xrange(165339, 173964), xrange(173967, 174233), xrange(176059, 176350), xrange(176355, 178042), xrange(178054, 178055), xrange(178064, 180039), xrange(180104, 180824), xrange(180847, 181491), xrange(181579, 182065), xrange(182069, 182963), xrange(182978,

We can delete the lists that we don't need anymore.

Now that we have the nucleotides, we can proceed with enhancing the feature set for our machine learning model. We are currently experimenting with various kind of feature sets. Their description and the corresponding code follows.

## Ancestor Transition Transversion Bits

In [33]:
del combinedList,AncestorNucleotide,AncestorTransitionTransversionSum,AncestorTransitionTransversionBits

## Without flanking regions

For this we only need to separate the combined list and get 2 lists one with indices and the other with labels.
The output of this would be 2 indices
1. IndicesList- Contains the indices
2. LabelList- Contains the labels

In [41]:
IndicesList = []
LabelList=[]
for i in range(len(combinedList)):
    if i%10000==0:
        print i
    IndicesList.append(combinedList[i][0:10])
    LabelList.append(combinedList[i][12:])

NameError: name 'combinedList' is not defined

In [ ]:
print len(IndicesList[0])
print len(LabelList[0])

Now we have the required indices and label list; so we can proceed and delete the combined list.

In [ ]:
del combinedList

You could also choose to have flanking regions along with the single column. The below few cells are used to get the flanking regions as well

## With flanking region

First decide on the number of flanking region on each side and set it below

<table><tr><th>Species</th><th>Class 0</th><th>Class 1</th><th>Class 2</th><th>Class 3</th><th>Class 4</th></tr><tr><td>Cow</td><td>40.076</td><td>19.257</td><td>9.142</td><td>23.82</td><td>7.706</td><td><tr><td>Cat</td><td>53.187</td><td>14.717</td><td>8.718</td><td>15.56</td><td>7.819</td><td><tr><td>Mouse</td><td>26.723</td><td>24.444</td><td>8.155</td><td>34.087</td><td>6.592</td><td><tr><td>Pig</td><td>50.141</td><td>15.821</td><td>9.012</td><td>17.561</td><td>7.464</td><td><tr><td>Rhesus</td><td>78.394</td><td>5.488</td><td>5.678</td><td>6.136</td><td>4.303</td><td><tr><td>Microbat</td><td>49.576</td><td>16.676</td><td>8.914</td><td>17.777</td><td>7.056</td><td><tr><td>Elephant</td><td>51.389</td><td>15.801</td><td>8.707</td><td>16.754</td><td>7.349</td><td><tr><td>Rabbit</td><td>50.267</td><td>15.603</td><td>9.037</td><td>16.402</td><td>8.69</td><td><tr><td>Shrew</td><td>24.88</td><td>24.961</td><td>8.156</td><td>35.899</td><td>6.105</td><td></table>

### Desampled Code

In [54]:
IndicesList = []
LabelList=[]
for i in range(len(DesampedFinal)):
    if i%1000000==0:
        print i
    if i>=noOfNeighborsOnOneSide and i<(len(DesampedFinal)-noOfNeighborsOnOneSide):
        tempList=[]
        flag =0
        for j in reversed(range(1,(noOfNeighborsOnOneSide+1))):
            if(DesampedFinal[i-j][10] == DesampedFinal[i-j+1][10]-1):
                tempList= tempList + DesampedFinal[i-j][0:10] + DesampedFinal[i-j][21:]
            else:
                flag = 1
        if(flag == 1):
            continue
        tempList+=DesampedFinal[i][0:10]
        for k in range(1,(noOfNeighborsOnOneSide+1)):
            if(DesampedFinal[i+k][10] == DesampedFinal[i+k-1][10]+1):
                tempList= tempList + DesampedFinal[i+k][0:10]+ DesampedFinal[i+k][21:]
            else:
                flag =1
        if(flag == 1):
            continue
        LabelList.append(DesampedFinal[i][12:21])
        IndicesList.append(tempList)

0
1000000
2000000
3000000
4000000


In [56]:
print LabelList[0]
print IndicesList[0]
print len(IndicesList)
print len(LabelList)

[0, 0, 3, 0, 0, 3, 3, 3, 1]
['G', 'G', 'G', '-', 'G', 'G', '-', '-', '-', '-', u'G', u'G', u'-', u'G', u'G', u'G', u'G', u'G', u'G', 0, 0, 14, 0, 4, 0, 'A', 'A', 'A', '-', 'A', 'A', '-', '-', '-', '-', u'A', u'A', u'-', u'A', u'A', u'A', u'A', u'A', u'A', 0, 0, 14, 0, 4, 0, 'T', 'T', 'T', '-', 'T', 'T', '-', '-', '-', '-', u'T', u'T', u'-', u'T', u'T', u'T', u'T', u'T', u'T', 0, 0, 14, 0, 4, 0, 'C', 'C', 'T', '-', 'C', 'C', '-', '-', '-', '-', 'T', 'T', 'T', '-', 'T', 'T', '-', '-', '-', '-', u'T', u'T', u'-', u'T', u'T', u'T', u'T', u'T', u'T', 0, 0, 14, 0, 4, 0, 'G', 'C', 'G', '-', 'G', 'G', '-', 'G', '-', '-', u'G', u'G', u'-', u'G', u'G', u'G', u'G', u'G', u'G', 0, 1, 14, 0, 3, 0, 'A', 'A', 'A', '-', 'A', 'A', '-', 'A', '-', '-', u'A', u'A', u'-', u'A', u'A', u'A', u'A', u'A', u'A', 0, 0, 15, 0, 3, 0]
3731533
3731533


In [57]:
get_species_wise_class_dist(LabelList)

'<table><tr><th>Species</th><th>Class 0</th><th>Class 1</th><th>Class 2</th><th>Class 3</th><th>Class 4</th></tr><tr><td>Cow</td><td>21.425</td><td>20.218</td><td>21.323</td><td>21.217</td><td>15.817</td><td><tr><td>Cat</td><td>43.145</td><td>16.411</td><td>16.359</td><td>15.85</td><td>8.235</td><td><tr><td>Mouse</td><td>20.279</td><td>25.588</td><td>14.242</td><td>33.297</td><td>6.594</td><td><tr><td>Pig</td><td>38.983</td><td>17.621</td><td>17.284</td><td>17.387</td><td>8.725</td><td><tr><td>Rhesus</td><td>70.327</td><td>7.946</td><td>10.927</td><td>5.329</td><td>5.471</td><td><tr><td>Microbat</td><td>39.451</td><td>18.168</td><td>16.538</td><td>18.121</td><td>7.722</td><td><tr><td>Elephant</td><td>41.576</td><td>17.817</td><td>15.755</td><td>17.784</td><td>7.068</td><td><tr><td>Rabbit</td><td>42.562</td><td>16.446</td><td>15.727</td><td>16.589</td><td>8.677</td><td><tr><td>Shrew</td><td>18.15</td><td>27.251</td><td>14.628</td><td>34.363</td><td>5.607</td><td></table>'

<table><tr><th>Species</th><th>Class 0</th><th>Class 1</th><th>Class 2</th><th>Class 3</th><th>Class 4</th></tr><tr><td>Cow</td><td>21.425</td><td>20.218</td><td>21.323</td><td>21.217</td><td>15.817</td><td><tr><td>Cat</td><td>43.145</td><td>16.411</td><td>16.359</td><td>15.85</td><td>8.235</td><td><tr><td>Mouse</td><td>20.279</td><td>25.588</td><td>14.242</td><td>33.297</td><td>6.594</td><td><tr><td>Pig</td><td>38.983</td><td>17.621</td><td>17.284</td><td>17.387</td><td>8.725</td><td><tr><td>Rhesus</td><td>70.327</td><td>7.946</td><td>10.927</td><td>5.329</td><td>5.471</td><td><tr><td>Microbat</td><td>39.451</td><td>18.168</td><td>16.538</td><td>18.121</td><td>7.722</td><td><tr><td>Elephant</td><td>41.576</td><td>17.817</td><td>15.755</td><td>17.784</td><td>7.068</td><td><tr><td>Rabbit</td><td>42.562</td><td>16.446</td><td>15.727</td><td>16.589</td><td>8.677</td><td><tr><td>Shrew</td><td>18.15</td><td>27.251</td><td>14.628</td><td>34.363</td><td>5.607</td><td></table>

In [58]:
del FinalCombinedList,combinedList,DesampedFinal

### Without Desampling

In [ ]:
IndicesList = []
LabelList=[]
for i in range(len(FinalCombinedList)):
    if i%1000000==0:
        print i
    if i>=noOfNeighborsOnOneSide and i<(len(FinalCombinedList)-noOfNeighborsOnOneSide):
        tempList=[]
        flag =0
        for j in reversed(range(1,(noOfNeighborsOnOneSide+1))):
            if(FinalCombinedList[i-j][10] == FinalCombinedList[i-j+1][10]-1):
                tempList= tempList + FinalCombinedList[i-j][0:10] + FinalCombinedList[i-j][21:]
            else:
                flag = 1
        if(flag == 1):
            continue
        tempList+=FinalCombinedList[i][0:10]
        for k in range(1,(noOfNeighborsOnOneSide+1)):
            if(FinalCombinedList[i+k][10] == FinalCombinedList[i+k-1][10]+1):
                tempList= tempList + FinalCombinedList[i+k][0:10]+ FinalCombinedList[i+k][21:]
            else:
                flag =1
        if(flag == 1):
            continue
        LabelList.append(FinalCombinedList[i][12:21])
        IndicesList.append(tempList)

In [ ]:
for i in FinalCombinedList[0:10]:
    print i

In [ ]:
print len(LabelList)
print len(DesampedFinal)

In [ ]:
assert(len(IndicesList[0])==(((noOfNeighborsOnOneSide*2)+1)*10))

In [ ]:
print IndicesList[0]
print FinalCombinedList[0:5]

If all is good we have our Label and Indices List with us. We can now delete the combinedList

In [ ]:
del combinedList

## Generating X and Y matrices

Based on what features you want to experiment with you can just add the different lists to get the X_training List. The Y_training List is just the LabelList. So we don't fiddle with that anymore.

In [70]:
#X_training=IndicesOneHotEncoded
#print len(X_training[0])
X_training=[]
X_training=[a + b for a, b in zip(IndicesOneHotEncoded, AncestorTransitionTransversionBits)]

In [72]:
print X_training[0]
print len(X_training[0])
print len(X_training)

[102, 102, 102, 104, 102, 102, 104, 104, 104, 104, 102, 102, 104, 102, 102, 102, 102, 102, 102, 0, 0, 14, 0, 4, 0, 100, 100, 100, 104, 100, 100, 104, 104, 104, 104, 100, 100, 104, 100, 100, 100, 100, 100, 100, 0, 0, 14, 0, 4, 0, 103, 103, 103, 104, 103, 103, 104, 104, 104, 104, 103, 103, 104, 103, 103, 103, 103, 103, 103, 0, 0, 14, 0, 4, 0, 101, 101, 103, 104, 101, 101, 104, 104, 104, 104, 103, 103, 103, 104, 103, 103, 104, 104, 104, 104, 103, 103, 104, 103, 103, 103, 103, 103, 103, 0, 0, 14, 0, 4, 0, 102, 101, 102, 104, 102, 102, 104, 102, 104, 104, 102, 102, 104, 102, 102, 102, 102, 102, 102, 0, 1, 14, 0, 3, 0, 100, 100, 100, 104, 100, 100, 104, 100, 104, 104, 100, 100, 104, 100, 100, 100, 100, 100, 100, 0, 0, 15, 0, 3, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0

In [73]:
#Y_training=[]
#for i in LabelList:
#    Y_training.append([i[0]])
Y_training = LabelList
#X_training = IndicesOneHotEncoded

In [74]:
print len(X_training[0])
print len(Y_training[0])
print Y_training[0]

268
9
[0, 0, 3, 0, 0, 3, 3, 3, 1]


****Only run for one output label type prediction (Eg: HC only)

In [83]:
y_pred = [[x] for x in y_pred]
y_train_pred = [[x] for x in y_train_pred]

## FNN

### Architecture 1

In [70]:
import tensorflow as tf
import keras
from random import random
from numpy import array
from numpy import cumsum
from keras.models import Sequential
from keras.layers import LSTM
from keras.layers import Dense
from keras.layers import TimeDistributed
from keras.layers import Bidirectional
import json
import numpy as np
tf.reset_default_graph()

In [78]:
from keras import backend as K
print K.learning_phase()
from keras.layers import Dropout
from keras import backend as K
from keras.datasets import imdb
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import LSTM
from keras.layers import Dropout
from keras.layers import Flatten
from keras.layers.embeddings import Embedding
from keras.preprocessing import sequence
from keras.metrics import categorical_accuracy as accuracy
from keras.objectives import sparse_categorical_crossentropy
from sklearn.model_selection import train_test_split
from sklearn.utils import class_weight
import tensorflow as tf

sess = tf.Session()
K.set_session(sess)

print len(X_train)
print len(y_train)
InputWidth = 160
OutputWidth = 9
#print Input[0:2]
#print Label[0:2]
inp = tf.placeholder(tf.float32, shape=(None,InputWidth))
out = tf.placeholder(tf.float32, shape=(None,OutputWidth))
print out.get_shape()
print inp.get_shape()
predictions=[]
predictionsTrain =[]
EmbeddingDimension = 5
OutputWidth = 10
learning_rate = 0.001
x = Embedding(1000, EmbeddingDimension, input_length=InputWidth)(inp)
print x
#x = K.reshape(x,(len(X_train)*InputWidth,5))
#print x
x = Flatten()(x)
print x
x = Dense(512)(x)
print x
x = Dense(256)(x)
print x
#To run for 1 output labels
#x = Dense(5)(x)
#print x
#preds = tf.reshape(x,((tf.shape(inp)[0],1,5)))
#print preds
#To run for 9 output labels
x = Dense(45)(x)
print x
preds = tf.reshape(x,((tf.shape(inp)[0],9,5)))
print preds

preds = Dense(5, activation='softmax')(preds)
print preds
#preds = K.argmax(preds,axis = -1)
#print preds
#acc_value = accuracy(labels, preds)
labels = tf.constant([[1, 2, 3,4,5]])
class_weights = tf.constant([[1.0, 2.0, 3.0,4.0,5.0]])
weights = tf.gather(class_weights, labels)

#loss = tf.reduce_mean(sparse_categorical_crossentropy(out, preds,weights))
loss = tf.reduce_mean(tf.losses.sparse_softmax_cross_entropy(labels=out, logits=preds,weights =weights))
#test_prediction = tf.nn.softmax(preds)
p = tf.argmax(preds, axis=2)
saver = tf.train.Saver()
#prediction=tf.argmax(y,2)
train_step = tf.train.AdamOptimizer(learning_rate=learning_rate).minimize(loss)
#acc_value = accuracy(out, preds)
with sess.as_default():
    sess.run(tf.global_variables_initializer())
    for epoch in range(10):
        print "epoch " + str(epoch)
    #train_step.run(feed_dict={inp:X_train,out:y_train,K.learning_phase(): 1})
        for i in range(10000):
            if(i%1000==0):
                print "iteration number"+str(i)
            train_step.run(feed_dict={inp:X_train[i*250:(i+1)*250],out:y_train[i*250:(i+1)*250],K.learning_phase(): 1})
        #temp=acc_value.eval(feed_dict={inp: X_train,out: y_train})
        #print type(temp)
        #print temp
        #save_path = saver.save(sess, "model_512_5neighbours_ExtraDense.ckpt")
        print out.get_shape()
        print preds.get_shape()
        #print preds.eval(feed_dict={inp:X_train[0:2],out: y_train[0:2]})
        p = tf.argmax(preds, axis=2)
        for i in range(10000):
            if(i%1000==0):
                print "iteration number"+str(i)
            predictions[i*100:(i+1)*100] = p.eval(feed_dict={inp: X_test[i*100:(i+1)*100],K.learning_phase(): 0})
            predictionsTrain[i*250:(i+1)*250] = p.eval(feed_dict={inp: X_train[i*250:(i+1)*250],K.learning_phase(): 0})
        print len(predictions)
        print len(predictionsTrain)
        print "train accuracy"
        get_accuracy(predictionsTrain,y_train[0:2500000])
        print "test accuracy"
        get_accuracy(predictions,y_test[0:1000000])
            
    #predictions =  p.eval(feed_dict={inp:X_train[0:2],out: y_train[0:2]})
    #print y_train[0:2]
    #print predictions
    #print precisiondictionOutput
    #predictions = sess.run(p,feed_dict={inp: X_test,K.learning_phase(): 0})
    #get_accuracy(predictions,y_test[0:300000])

    
#p = tf.argmax(preds, axis=2)
#print p.get_shape()
#print p.eval(feed_dict={inp: X_train,out: y_train})
'''
pout = K.reshape(out,((len(X_test)*9)/9,9))
print pout.eval(feed_dict={inp: X_test,out: y_test})
print out.eval(feed_dict={inp: X_test,out: y_test})
acc_value = accuracy(pout, p)
print acc_value
#print acc_value.eval()
print acc_value.eval(feed_dict={inp: X_test,out: y_test,K.learning_phase(): 0})
save_path = saver.save(sess, "model.ckpt")
'''
#Label = np.asarray(Label)

Tensor("keras_learning_phase:0", dtype=bool)
2521616
2521616
(?, 9)
(?, 160)
Tensor("embedding_7/Gather:0", shape=(?, 160, 5), dtype=float32)
Tensor("flatten_7/Reshape:0", shape=(?, ?), dtype=float32)
Tensor("dense_25/BiasAdd:0", shape=(?, 512), dtype=float32)
Tensor("dense_26/BiasAdd:0", shape=(?, 256), dtype=float32)
Tensor("dense_27/BiasAdd:0", shape=(?, 45), dtype=float32)
Tensor("Reshape_6:0", shape=(?, 9, 5), dtype=float32)
Tensor("dense_28/div:0", shape=(?, 9, 5), dtype=float32)


ValueError: Can not squeeze dim[2], expected a dimension of 1, got 5 for 'sparse_softmax_cross_entropy_loss/Squeeze' (op: 'Squeeze') with input shapes: [1,5,5].

In [69]:
print "Training Accuracy"
get_accuracy_SpeciesWise(predictionsTrain,y_train[0:2500000],"Cow")
print "Test Accuracy"
get_accuracy_SpeciesWise(predictions,y_test[0:1000000],"Cow")

Training Accuracy
Overall Accuracy is 0.5363832
Class 0: 387754 0 103094 0 64249 555097 0.698533769774
Class 1: 0 217681 0 305325 0 523006 0.416211286295
Class 2: 171012 0 235074 0 63471 469557 0.500629316569
Class 3: 0 146828 0 402552 0 549380 0.732738723652
Class 4: 210709 0 94354 0 97897 402960 0.242944709152
Test Accuracy
Overall Accuracy is 0.536298
Class 0: 155150 0 41375 0 25336 221861 0.699311731219
Class 1: 0 87131 0 122608 0 209739 0.415425838781
Class 2: 68467 0 93983 0 25524 187974 0.499978720461
Class 3: 0 58753 0 161068 0 219821 0.732723443165
Class 4: 84018 0 37621 0 38966 160605 0.242620092774


### FNN Architecture 2

In [ ]:
from keras import backend as K
print K.learning_phase()
from keras.layers import Dropout
from keras import backend as K
from keras.datasets import imdb
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import LSTM
from keras.layers import Dropout
from keras.layers import Flatten
from keras.layers.embeddings import Embedding
from keras.preprocessing import sequence
from keras.layers import Bidirectional
from keras.metrics import categorical_accuracy as accuracy
from keras.objectives import sparse_categorical_crossentropy
from sklearn.model_selection import train_test_split

sess = tf.Session()
K.set_session(sess)
#Input = Input[0:2]
#Label = Label[0:2]
#Input =[]
#Label=[]
#Input = [[[1],[2],[2],[4],[5],[4],[3],[4],[5],[5]],[[1],[2],[2],[2],[5],[3],[3],[4],[5],[5]]]
#Label = [[[1],[0],[1],[2],[1],[0],[4],[4],[2],[1]],[[2],[0],[1],[2],[1],[0],[4],[4],[2],[1]]]
#Input = [[1,2,3,4,5,4,5,3,2,1],[1,2,3,4,5,4,3,2,1,2]]
#Label = [[1,2,3,4,0,4,0,3,2,1],[1,2,3,4,0,4,3,2,1,2]]
#TInput = [[1,2,3,4,5,4,5,3,2,1],[1,2,3,4,5,4,3,2,1,2]]
#TLabel = [[1,2,3,4,0,4,0,3,2,1],[1,2,3,4,0,4,3,2,1,2]]
'''
Input = [[[1,0,0,0,0],[0,1,0,0,0],[0,1,0,0,0],[0,0,0,1,0],[0,0,0,0,1],[0,0,0,1,0],[0,0,1,0,0],[0,0,0,1,0],[0,0,0,0,1],
          [0,0,0,0,1]],
         [[1,0,0,0,0],[0,1,0,0,0],[0,1,0,0,0],[0,1,0,0,0],[0,0,0,0,1],[0,0,1,0,0],[0,0,1,0,0],[0,0,0,1,0],[0,0,0,0,1],
          [0,0,0,0,1]]]

Label = [[[1,0,0,0,0],[1,0,0,0,0],[0,1,0,0,0],[0,1,0,0,0],[0,0,0,1,0],[0,0,0,0,1],[0,0,0,1,0],[0,0,1,0,0],[0,0,0,1,0],[0,0,0,0,1]],
         [[1,0,0,0,0],[1,0,0,0,0],[0,1,0,0,0],[0,1,0,0,0],[0,1,0,0,0],[0,0,0,0,1],[0,0,1,0,0],[0,0,1,0,0],[0,0,0,1,0],[0,0,0,0,1]]
         ]
'''



#print Input[0:2]
#print Label[0:2]
inp = tf.placeholder(tf.float32, shape=(None,250))
out = tf.placeholder(tf.float32, shape=(None,9))
print out.get_shape()
print inp.get_shape()
t = inp.get_shape().as_list()
print t[0]
EmbeddingDimension = 6
InputWidth = 250
learning_rate=0.001
predictions =[]
predictionsTrain =[]
x = Embedding(10, EmbeddingDimension, input_length=InputWidth)(inp)
print x
x = Flatten()(x)
print x
x = Dense(512)(x)
print x
x = Dense(256)(x)
print x
#shape = K.shape(x)
#pool_shape = tf.stack([t[0]*InputWidth,200])  # Here you can mix integers and symbolic elements of `shape`
#x = tf.reshape(x, pool_shape)
#x = Dropout(0.5)(x)
#print x
x = Dense(45)(x)
print x
#preds = x
preds = tf.reshape(x,((tf.shape(inp)[0],9,5)))
print preds
preds = Dense(5,activation='softmax')(preds)
print preds
#preds = K.argmax(preds,axis = -1)
#print preds

loss = tf.reduce_mean(sparse_categorical_crossentropy(out, preds))
#test_prediction = tf.nn.softmax(preds)
p = tf.argmax(preds, axis=2)
train_step = tf.train.AdamOptimizer(learning_rate=learning_rate).minimize(loss)
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    for epoch in range(10):
    #train_step.run(feed_dict={inp:X_train,out:y_train,K.learning_phase(): 1})
        for i in range(10000):
            if(i%1000==0):
                print "iteration number"+str(i)
            train_step.run(feed_dict={inp:X_train[i*500:(i+1)*500],out:y_train[i*500:(i+1)*500],K.learning_phase(): 1})
        #temp=acc_value.eval(feed_dict={inp: X_train,out: y_train})
        #print type(temp)
        #print temp
        #save_path = saver.save(sess, "model_512_5neighbours_ExtraDense.ckpt")
        print out.get_shape()
        print preds.get_shape()
        #print preds.eval(feed_dict={inp:X_train[0:2],out: y_train[0:2]})
        p = tf.argmax(preds, axis=2)
        for i in range(1000):
            if(i%100==0):
                print "iteration number"+str(i)
            predictions[i*300:(i+1)*300] = p.eval(feed_dict={inp: X_test[i*300:(i+1)*300],K.learning_phase(): 0})
            predictionsTrain[i*300:(i+1)*300] = p.eval(feed_dict={inp: X_train[i*300:(i+1)*300],K.learning_phase(): 0})
    #predictions =  p.eval(feed_dict={inp:X_train[0:2],out: y_train[0:2]})
    #print y_train[0:2]
    #print predictions
    #print precisiondictionOutput
    #predictions = sess.run(p,feed_dict={inp: X_test,K.learning_phase(): 0})
    #get_accuracy(predictions,y_test[0:300000])
#Label = np.asarray(Label)


